In [8]:
import numpy as np
import pandas as pd

In [9]:

df = pd.read_csv("smartphones_smartphones.csv")

df.head()

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
0,OnePlus 11 5G,"₹54,999",89.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor","12 GB RAM, 256 GB inbuilt",5000 mAh Battery with 100W Fast Charging,"6.7 inches, 1440 x 3216 px, 120 Hz Display wit...",50 MP + 48 MP + 32 MP Triple Rear & 16 MP Fron...,Memory Card Not Supported,Android v13
1,OnePlus Nord CE 2 Lite 5G,"₹19,989",81.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 33W Fast Charging,"6.59 inches, 1080 x 2412 px, 120 Hz Display wi...",64 MP + 2 MP + 2 MP Triple Rear & 16 MP Front ...,"Memory Card (Hybrid), upto 1 TB",Android v12
2,Samsung Galaxy A14 5G,"₹16,499",75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Exynos 1330, Octa Core, 2.4 GHz Processor","4 GB RAM, 64 GB inbuilt",5000 mAh Battery with 15W Fast Charging,"6.6 inches, 1080 x 2408 px, 90 Hz Display with...",50 MP + 2 MP + 2 MP Triple Rear & 13 MP Front ...,"Memory Card Supported, upto 1 TB",Android v13
3,Motorola Moto G62 5G,"₹14,999",81.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with Fast Charging,"6.55 inches, 1080 x 2400 px, 120 Hz Display wi...",50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,"Memory Card (Hybrid), upto 1 TB",Android v12
4,Realme 10 Pro Plus,"₹24,999",82.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Dimensity 1080, Octa Core, 2.6 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 67W Fast Charging,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit...",108 MP + 8 MP + 2 MP Triple Rear & 16 MP Front...,Memory Card Not Supported,Android v13


2 types of Data Cleaning ideologies:

1. Noisy Data/ Dirty Data- Useless data, missing values.
2. Messy data - Useful Data but unstructured. --> Structure

### Major types of data cleanliness issues:


1. **Completeness Issues** – Data is missing or incomplete where it is expected to be present.

2. **Validity Issues** – Data doesn't conform to defined formats, types, or rules (e.g., age = -5).

3. **Consistency Issues** – Conflicting data exists across different systems or columns (e.g., birthdate vs. age mismatch).

4. **Accuracy Issues** – Data does not reflect the real-world values correctly (e.g., wrong addresses or names).

5. **Uniqueness Issues** – Duplicate or repeated entries exist where only one should be present.

---


In [12]:
# shape of the data

df.shape

(1020, 11)

In [13]:
# check for missing values

df.isna().sum()

model          0
price          0
rating       141
sim            0
processor      0
ram            0
battery        0
display        0
camera         1
card           7
os            17
dtype: int64

In [14]:
(
    df
    .isna()
    .any(axis=1)
)

0       False
1       False
2       False
3       False
4       False
        ...  
1015    False
1016    False
1017    False
1018    False
1019    False
Length: 1020, dtype: bool

In [15]:
# check for rows having missing values

df.loc[(
    df
    .isna()
    .any(axis=1)
),:]

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
14,Samsung Galaxy S23 Ultra 5G,"₹1,14,990",NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC","Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor","8 GB RAM, 256 GB inbuilt",5000 mAh Battery with 45W Fast Charging,"6.8 inches, 1440 x 3088 px, 120 Hz Display wit...",200 MP Quad Rear & 12 MP Front Camera,Memory Card Not Supported,Android v13
29,OnePlus 11 Pro,"₹69,999",NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor","8 GB RAM, 128 GB inbuilt",4500 mAh Battery with 100W Fast Charging,"6.7 inches, 1440 x 3216 px, 120 Hz Display wit...",50 MP + 48 MP + 32 MP Triple Rear & 16 MP Fron...,Android v13,No FM Radio
37,Samsung Galaxy S22 Ultra 5G,"₹91,999",NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8 Gen1, Octa Core, 3 GHz Processor","12 GB RAM, 256 GB inbuilt",5000 mAh Battery with 45W Fast Charging,"6.8 inches, 1440 x 3080 px, 120 Hz Display wit...",108 MP Quad Rear & 40 MP Front Camera,Android v12,Bluetooth
49,Samsung Galaxy A74 5G,"₹42,999",NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 7 Gen1, Octa Core, 2.36 GHz Processor","8 GB RAM, 128 GB inbuilt",5000 mAh Battery with 33W Fast Charging,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit...",108 MP Quad Rear & 32 MP Front Camera,"Memory Card (Hybrid), upto 1 TB",Android v12
69,Oppo Find N Fold,"₹99,990",NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8+ Gen1, Octa Core, 3.2 GHz Processor","8 GB RAM, 256 GB inbuilt",5000 mAh Battery with 67W Fast Charging,"7.1 inches, 1792 x 1920 px, 120 Hz Display wit...","Foldable Display, Dual Display",64 MP + 10 MP + 8 MP Triple Rear & 32 MP Front...,Android v12
...,...,...,...,...,...,...,...,...,...,...,...
996,Vivo Y55S,"₹13,490",NaN,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Snapdragon 425, Quad Core, 1.2 GHz Processor","3 GB RAM, 16 GB inbuilt",2730 mAh Battery,"5.2 inches, 720 x 1280 px Display",13 MP Rear & 5 MP Front Camera,Memory Card Supported,Android v6.0 (Marshmallow)
1000,XTouch F40 Flip,"₹1,999",NaN,Dual Sim,No 3G,No Wifi,"32 MB RAM, 32 MB inbuilt",800 mAh Battery,"1.77 inches, 240 x 320 px Display",Dual Display,1.3 MP Rear Camera
1003,Lava X3,"₹6,999",NaN,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Helio A22, Quad Core, 2 GHz Processor","3 GB RAM, 32 GB inbuilt",4000 mAh Battery,"6.5 inches, 1600 x 720 px Display with Water D...",8 MP Dual Rear & 5 MP Front Camera,"Memory Card (Hybrid), upto 512 GB",Android v12
1012,itel A23s,"₹4,787",NaN,"Dual Sim, 3G, 4G, Wi-Fi","Spreadtrum SC9832E, Quad Core, 1.4 GHz Processor","2 GB RAM, 32 GB inbuilt",3020 mAh Battery,"5 inches, 854 x 480 px Display",2 MP Rear Camera,Android v11,No FM Radio


In [16]:
# duplicate data

df.duplicated().sum()

0

## Overall Observations

- There are around 1020 rows and 11 columns in the data.
- There are no duplicate rows in the data.
- There are a few missing values in the data in columns:
    1. `rating`
    2. `camera`
    3. `card`
    4. `os`
- Remove feature phones from the data.

In [18]:
df.columns

Index(['model', 'price', 'rating', 'sim', 'processor', 'ram', 'battery',
       'display', 'camera', 'card', 'os'],
      dtype='object')

## Column Wise Observations

### Model

- The model column has two pieces of info.
    1. Brand Name
    2. Model Name
- Row 15, 45, 50, 52 We have Ram and storage(in paranthesis) info along with model info - `Validation`

### Price
- Rupee sign - `Validation`
- Commas in values - `Validation`

### Rating
- Missing values in column - 141 - `Completeness`
- Ratings should be on the scale of 10 - `Consistency`

### Sim
- We have to see the number of SIM's supported.
- Bands Supported
- Has Wifi
- Has Extra features - (NFC, IR Blaster)

### Processor
- Processor name
    1. Brand Name
    2. Model Name
- Number of Cores - Str --> Integers
- CPU speed given in GHz

#### Row wise
    - Row 120- tesla phone - Here only number of cores are given - `accuracy`
    - Row 154 - Only Processor name is given - `consistency`
    - Row 203 - Processor name is not given - `consistency`
    - Row 584 - Data validation and inaccurate

### RAM
- The first piece of info is about RAM.(in GB)
- The second is about storage.(in GB)

- Apple phone has only storage info.

### Battery
1. Size of the battery.
2. Fast charging present or not.
3. If FC then wattage.

### Display
1. Display size.
2. Resolution
3. Refresh Rate
4. Display shape

- Apple phones have data shifting issues.

### Camera
1. Number of cameras.
2. Resolution of back cameras.
3. Resolution of front camera.

### Card
1. Is supported.
2. Sim Tray config
3. Size

- Some rows have OS info in the memory card column.

### OS
1. OS --> Android iOS
2. Version


**Feature Phones**

1. Low priced. Lower than Rs.5000/-.
2. Small screen size - Less than 3 inch
3. Smaller battery capicities - Lower than 2000 mAh

In [21]:
df.sample(30)

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
917,Nokia 150 (2020),"₹2,499",NaN,Dual Sim,No Wifi,4 MB inbuilt,1020 mAh Battery,"2.4 inches, 240 x 320 px Display",0.3 MP Rear Camera,"Memory Card Supported, upto 32 GB",Bluetooth
827,Vivo X80 5G (12GB RAM + 256GB),"₹59,999",88.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Dimensity 9000, Octa Core, 3.05 GHz Processor","12 GB RAM, 256 GB inbuilt",4500 mAh Battery with 80W Fast Charging,"6.78 inches, 1080 x 2400 px, 120 Hz Display wi...",50 MP + 12 MP + 12 MP Triple Rear & 32 MP Fron...,Memory Card Not Supported,Android v12
337,Samsung Galaxy A03 Core,"₹6,499",NaN,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Unisoc SC9836A, Quad Core, 1.6 GHz Processor","2 GB RAM, 32 GB inbuilt",5000 mAh Battery,"6.5 inches, 720 x 1600 px Display with Water D...",8 MP Rear & 5 MP Front Camera,"Memory Card Supported, upto 1 TB",Android v10
643,Xiaomi 11T Pro 5G (12GB RAM + 256GB),"₹38,999",88.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...","Snapdragon 888 , Octa Core, 2.84 GHz Processor","12 GB RAM, 256 GB inbuilt",5000 mAh Battery with 120W Fast Charging,"6.67 inches, 1080 x 2400 px, 120 Hz Display wi...",108 MP + 8 MP + 5 MP Triple Rear & 16 MP Front...,Android v11,No FM Radio
656,Apple iPhone SE 2020,"₹39,900",63.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi, NFC","A13 Bionic, Hexa Core, 2.65 GHz Processor","3 GB RAM, 64 GB inbuilt",1821 mAh Battery with 18W Fast Charging,"4.7 inches, 750 x 1334 px Display",12 MP Rear & 7 MP Front Camera,iOS v13,No FM Radio
385,iQOO Z6 5G (8GB RAM + 128GB),"₹18,498",82.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 695 , Octa Core, 2.2 GHz Processor","8 GB RAM, 128 GB inbuilt",5000 mAh Battery with 18W Fast Charging,"6.58 inches, 1080 x 2408 px, 120 Hz Display wi...",50 MP + 2 MP + 2 MP Triple Rear & 16 MP Front ...,"Memory Card (Hybrid), upto 1 TB",Android v12
1,OnePlus Nord CE 2 Lite 5G,"₹19,989",81.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 33W Fast Charging,"6.59 inches, 1080 x 2412 px, 120 Hz Display wi...",64 MP + 2 MP + 2 MP Triple Rear & 16 MP Front ...,"Memory Card (Hybrid), upto 1 TB",Android v12
266,OPPO Reno7 Pro 5G,"₹31,994",87.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Dimensity 1200 Max, Octa Core, 3 GHz Processor","12 GB RAM, 256 GB inbuilt",4500 mAh Battery with 65W Fast Charging,"6.55 inches, 1080 x 2400 px, 90 Hz Display wit...",50 MP + 8 MP + 2 MP Triple Rear & 32 MP Front ...,Memory Card Supported,Android v11
739,Realme Q3,"₹14,999",80.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 750G , Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 30W Fast Charging,"6.5 inches, 1080 x 2400 px, 120 Hz Display wit...",48 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,Memory Card Supported,Android v11
180,POCO M4 Pro 4G,"₹11,999",77.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi, IR Blaster","Helio G96, Octa Core, 2.05 GHz Processor","6 GB RAM, 64 GB inbuilt",5000 mAh Battery with 33W Fast Charging,"6.43 inches, 1080 x 2400 px, 90 Hz Display wit...",64 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,"Memory Card Supported, upto 1 TB",Android v11


In [22]:
# remove feature phones from data

def extract_feature_phones(df: pd.DataFrame):
    temp_df = (
        df
        .assign(
            price=lambda df_: (
                df_["price"]
                .str.replace("₹", "")
                .str.replace(",", "")
                .astype(np.float64)
            ),
            battery = lambda df_: (
                df_["battery"]
                .str.extract("(\d+).?mAh")
                .astype(np.float64)
        ),
            display = lambda df_:(
                df_["display"]
                .str.extract("(\d+\.\d*).?inches")
                .astype(np.float64)
            )
        )
    )

    battery_cond = temp_df["battery"] <= 2000
    display_cond = temp_df["display"] <= 4.0
    price_cond = temp_df["price"] <= 4000


    return (
        temp_df
        .loc[(battery_cond) | (display_cond) | (price_cond)]
    )


extract_feature_phones(df)

<>:15: SyntaxWarning: invalid escape sequence '\d'
<>:20: SyntaxWarning: invalid escape sequence '\d'
<>:15: SyntaxWarning: invalid escape sequence '\d'
<>:20: SyntaxWarning: invalid escape sequence '\d'
C:\Users\dipak\AppData\Local\Temp\ipykernel_18756\39041887.py:15: SyntaxWarning: invalid escape sequence '\d'
  .str.extract("(\d+).?mAh")
C:\Users\dipak\AppData\Local\Temp\ipykernel_18756\39041887.py:20: SyntaxWarning: invalid escape sequence '\d'
  .str.extract("(\d+\.\d*).?inches")


,model,price,rating,sim,processor,ram,battery,display,camera,card,os
155,Nokia 2780 Flip,4990.0,NaN,"Dual Sim, 3G, 4G, Wi-Fi","Snapdragon QM215, Quad Core, 1.3 GHz Processor","4 GB RAM, 512 MB inbuilt",1450.0,2.70,Dual Display,5 MP Rear Camera,"Memory Card Supported, upto 32 GB"
190,Realme C2s,3499.0,63.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Helio P22 , Octa Core, 2 GHz Processor","3 GB RAM, 32 GB inbuilt",4000.0,6.10,13 MP + 2 MP Dual Rear & 5 MP Front Camera,"Memory Card Supported, upto 128 GB",Android v9.0 (Pie)
191,Duoqin F22 Pro,9990.0,NaN,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Helio G85, Octa Core, 2 GHz Processor","4 GB RAM, 64 GB inbuilt",2150.0,3.54,8 MP Rear & 2 MP Front Camera,Memory Card Supported,Android v12
271,Nokia 2720 V Flip,6199.0,NaN,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Snapdragon 205 , Dual Core, 1.1 GHz Processor","512 MB RAM, 4 GB inbuilt",1500.0,2.80,Dual Display,2 MP Rear Camera,Memory Card Supported
400,Jio JioPhone 2,2999.0,NaN,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Dual Core, 1 GHz Processor","512 MB RAM, 4 GB inbuilt",2000.0,2.40,2 MP Rear & 0.3 MP Front Camera,"Memory Card Supported, upto 128 GB",KAI OS
473,Nokia 110 4G,1762.0,NaN,"Dual Sim, 3G, 4G, VoLTE",No Wifi,"128 MB RAM, 48 MB inbuilt",1020.0,1.80,0.3 MP Rear Camera,"Memory Card Supported, upto 32 GB",NaN
477,Vivo Y25,7499.0,NaN,"Dual Sim, 3G, 4G, Wi-Fi","Helio MT6580, Quad Core, 1.3 GHz Processor","1 GB RAM, 16 GB inbuilt",1900.0,4.50,5 MP Rear & 2 MP Front Camera,"Memory Card Supported, upto 128 GB",Android v5.1 (Lollipop)
478,Lyf Earth 1,3990.0,60.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Snapdragon 615, Octa Core, 1.5 GHz Processor","3 GB RAM, 32 GB inbuilt",3500.0,5.50,13 MP + 2 MP Dual Rear & 5 MP Front Camera,"Memory Card (Hybrid), upto 64 GB",Android v5.1.1 (Lollipop)
486,Nokia 8000 4G,6899.0,NaN,"Dual Sim, 3G, 4G, Wi-Fi","Snapdragon 210, Quad Core, 1.1 GHz Processor","512 MB RAM, 4 GB inbuilt",1500.0,2.80,2 MP Rear Camera,Memory Card Supported,Bluetooth
532,Samsung Guru Music 2 Dual Sim,1949.0,NaN,Dual Sim,No Wifi,"Single Core, 208 MHz Processor",800.0,NaN,No Rear Camera,"Memory Card Supported, upto 16 GB",NaN


In [23]:
df

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
0,OnePlus 11 5G,"₹54,999",89.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor","12 GB RAM, 256 GB inbuilt",5000 mAh Battery with 100W Fast Charging,"6.7 inches, 1440 x 3216 px, 120 Hz Display wit...",50 MP + 48 MP + 32 MP Triple Rear & 16 MP Fron...,Memory Card Not Supported,Android v13
1,OnePlus Nord CE 2 Lite 5G,"₹19,989",81.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 33W Fast Charging,"6.59 inches, 1080 x 2412 px, 120 Hz Display wi...",64 MP + 2 MP + 2 MP Triple Rear & 16 MP Front ...,"Memory Card (Hybrid), upto 1 TB",Android v12
2,Samsung Galaxy A14 5G,"₹16,499",75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Exynos 1330, Octa Core, 2.4 GHz Processor","4 GB RAM, 64 GB inbuilt",5000 mAh Battery with 15W Fast Charging,"6.6 inches, 1080 x 2408 px, 90 Hz Display with...",50 MP + 2 MP + 2 MP Triple Rear & 13 MP Front ...,"Memory Card Supported, upto 1 TB",Android v13
3,Motorola Moto G62 5G,"₹14,999",81.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with Fast Charging,"6.55 inches, 1080 x 2400 px, 120 Hz Display wi...",50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,"Memory Card (Hybrid), upto 1 TB",Android v12
4,Realme 10 Pro Plus,"₹24,999",82.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Dimensity 1080, Octa Core, 2.6 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 67W Fast Charging,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit...",108 MP + 8 MP + 2 MP Triple Rear & 16 MP Front...,Memory Card Not Supported,Android v13
...,...,...,...,...,...,...,...,...,...,...,...
1015,Motorola Moto Edge S30 Pro,"₹34,990",83.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 8 Gen1, Octa Core, 3 GHz Processor","8 GB RAM, 128 GB inbuilt",5000 mAh Battery with 68.2W Fast Charging,"6.67 inches, 1080 x 2460 px, 120 Hz Display wi...",64 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,Android v12,No FM Radio
1016,Honor X8 5G,"₹14,990",75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 480+, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 22.5W Fast Charging,"6.5 inches, 720 x 1600 px Display with Water D...",48 MP + 2 MP + Depth Sensor Triple Rear & 8 MP...,"Memory Card Supported, upto 1 TB",Android v11
1017,POCO X4 GT 5G (8GB RAM + 256GB),"₹28,990",85.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...","Dimensity 8100, Octa Core, 2.85 GHz Processor","8 GB RAM, 256 GB inbuilt",5080 mAh Battery with 67W Fast Charging,"6.6 inches, 1080 x 2460 px, 144 Hz Display wit...",64 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,Memory Card Not Supported,Android v12
1018,Motorola Moto G91 5G,"₹19,990",80.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with Fast Charging,"6.8 inches, 1080 x 2400 px Display with Punch ...",108 MP + 8 MP + 2 MP Triple Rear & 32 MP Front...,"Memory Card Supported, upto 1 TB",Android v12


In [24]:
feature_phones_index = extract_feature_phones(df).index.to_list()

In [25]:
print(feature_phones_index)
print(len(feature_phones_index))

[155, 190, 191, 271, 400, 473, 477, 478, 486, 532, 533, 551, 573, 575, 582, 608, 611, 627, 640, 647, 656, 657, 665, 685, 699, 748, 757, 817, 835, 844, 846, 852, 873, 881, 907, 917, 925, 926, 927, 930, 932, 988, 1000]
43


In [26]:
def incomplete_data_phones(data: pd.DataFrame):

    columns_list = ["processor", "ram", "display"]

    temp_df = (
        data
        .assign(
            **{
                col: data[col].str.count(",")
                for col in columns_list
            }
        )
    )

    processor_split = 2
    ram_split = 1
    display_split = 2

    return (
        data
        .loc[(temp_df["processor"].lt(processor_split)) | \
        (temp_df["ram"].lt(ram_split)) | \
        (temp_df["display"].lt(display_split))]
    )



incomplete_data_phones(df)

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
6,Apple iPhone 14,"₹65,999",81.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A15, Hexa Core, 3.22 GHz Processor","6 GB RAM, 128 GB inbuilt",3279 mAh Battery with Fast Charging,"6.1 inches, 1170 x 2532 px Display with Small ...",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v16
15,Apple iPhone 13,"₹62,999",79.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A15, Hexa Core, 3.22 GHz Processor","4 GB RAM, 128 GB inbuilt",3240 mAh Battery with Fast Charging,"6.1 inches, 1170 x 2532 px Display with Small ...",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v15
16,Vivo Y16,"₹9,999",65.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Helio P35, Octa Core, 2.3 GHz Processor","3 GB RAM, 32 GB inbuilt",5000 mAh Battery with 10W Fast Charging,"6.51 inches, 720 x 1600 px Display with Water ...",13 MP + 2 MP Dual Rear & 5 MP Front Camera,Memory Card Supported,Android v12
19,Vivo Y22,"₹14,499",72.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Helio G70, Octa Core, 2 GHz Processor","4 GB RAM, 64 GB inbuilt",5000 mAh Battery with 18W Fast Charging,"6.55 inches, 720 x 1612 px Display with Water ...",50 MP + 2 MP Dual Rear & 8 MP Front Camera,"Memory Card Supported, upto 1 TB",Android v12
27,Apple iPhone 14 Pro Max,"₹1,29,990",76.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A16, Hexa Core Processor","6 GB RAM, 128 GB inbuilt",4323 mAh Battery with Fast Charging,"6.7 inches, 1290 x 2796 px, 120 Hz Display",48 MP + 12 MP + 12 MP Triple Rear & 12 MP Fron...,Memory Card Not Supported,iOS v16
...,...,...,...,...,...,...,...,...,...,...,...
1012,itel A23s,"₹4,787",NaN,"Dual Sim, 3G, 4G, Wi-Fi","Spreadtrum SC9832E, Quad Core, 1.4 GHz Processor","2 GB RAM, 32 GB inbuilt",3020 mAh Battery,"5 inches, 854 x 480 px Display",2 MP Rear Camera,Android v11,No FM Radio
1013,Google Pixel 8 Pro,"₹70,990",80.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Google Tensor 3, Octa Core Processor","12 GB RAM, 256 GB inbuilt",5000 mAh Battery with 67W Fast Charging,"6.73 inches, 1440 x 3120 px, 120 Hz Display wi...",50 MP + 50 MP + 50 MP Triple Rear & 12 MP Fron...,Android v13,No FM Radio
1016,Honor X8 5G,"₹14,990",75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 480+, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 22.5W Fast Charging,"6.5 inches, 720 x 1600 px Display with Water D...",48 MP + 2 MP + Depth Sensor Triple Rear & 8 MP...,"Memory Card Supported, upto 1 TB",Android v11
1018,Motorola Moto G91 5G,"₹19,990",80.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with Fast Charging,"6.8 inches, 1080 x 2400 px Display with Punch ...",108 MP + 8 MP + 2 MP Triple Rear & 32 MP Front...,"Memory Card Supported, upto 1 TB",Android v12


In [27]:
df.head(3)

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
0,OnePlus 11 5G,"₹54,999",89.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor","12 GB RAM, 256 GB inbuilt",5000 mAh Battery with 100W Fast Charging,"6.7 inches, 1440 x 3216 px, 120 Hz Display wit...",50 MP + 48 MP + 32 MP Triple Rear & 16 MP Fron...,Memory Card Not Supported,Android v13
1,OnePlus Nord CE 2 Lite 5G,"₹19,989",81.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 33W Fast Charging,"6.59 inches, 1080 x 2412 px, 120 Hz Display wi...",64 MP + 2 MP + 2 MP Triple Rear & 16 MP Front ...,"Memory Card (Hybrid), upto 1 TB",Android v12
2,Samsung Galaxy A14 5G,"₹16,499",75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Exynos 1330, Octa Core, 2.4 GHz Processor","4 GB RAM, 64 GB inbuilt",5000 mAh Battery with 15W Fast Charging,"6.6 inches, 1080 x 2408 px, 90 Hz Display with...",50 MP + 2 MP + 2 MP Triple Rear & 13 MP Front ...,"Memory Card Supported, upto 1 TB",Android v13


In [28]:
import re

In [29]:
def get_apple_phones(data: pd.DataFrame):
    return (
        data
        .loc[(
            data["model"]
            .str.contains("apple", flags=re.IGNORECASE)
        ), :]
    )

get_apple_phones(df)

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
6,Apple iPhone 14,"₹65,999",81.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A15, Hexa Core, 3.22 GHz Processor","6 GB RAM, 128 GB inbuilt",3279 mAh Battery with Fast Charging,"6.1 inches, 1170 x 2532 px Display with Small ...",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v16
15,Apple iPhone 13,"₹62,999",79.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A15, Hexa Core, 3.22 GHz Processor","4 GB RAM, 128 GB inbuilt",3240 mAh Battery with Fast Charging,"6.1 inches, 1170 x 2532 px Display with Small ...",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v15
27,Apple iPhone 14 Pro Max,"₹1,29,990",76.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A16, Hexa Core Processor","6 GB RAM, 128 GB inbuilt",4323 mAh Battery with Fast Charging,"6.7 inches, 1290 x 2796 px, 120 Hz Display",48 MP + 12 MP + 12 MP Triple Rear & 12 MP Fron...,Memory Card Not Supported,iOS v16
33,Apple iPhone 11,"₹38,999",73.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi, NFC","A13 Bionic, Hexa Core, 2.65 GHz Processor","4 GB RAM, 64 GB inbuilt",3110 mAh Battery,"6.1 inches, 828 x 1792 px Display with Large N...",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v13
56,Apple iPhone 14 Plus,"₹74,999",82.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A15, Hexa Core, 3.22 GHz Processor","6 GB RAM, 128 GB inbuilt",4325 mAh Battery with Fast Charging,"6.7 inches, 1284 x 2778 px Display with Small ...",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v16
76,Apple iPhone 13 Pro,"₹1,19,900",83.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A15, Hexa Core, 3.22 GHz Processor","6 GB RAM, 128 GB inbuilt",3095 mAh Battery with Fast Charging,"6.1 inches, 1170 x 2532 px, 120 Hz Display wit...",12 MP + 12 MP + 12 MP Triple Rear & 12 MP Fron...,Memory Card Not Supported,iOS v15.0
100,Apple iPhone 14 Pro,"₹1,19,990",75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A16, Hexa Core Processor","6 GB RAM, 128 GB inbuilt",3200 mAh Battery with Fast Charging,"6.1 inches, 1179 x 2556 px, 120 Hz Display",48 MP + 12 MP + 12 MP Triple Rear & 12 MP Fron...,Memory Card Not Supported,iOS v16
111,Apple iPhone 12,"₹51,999",74.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 64 GB inbuilt","6.1 inches, 1170 x 2532 px Display with Large ...",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14,No FM Radio
136,Apple iPhone 11 (128GB),"₹46,999",75.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi, NFC","A13 Bionic, Hexa Core, 2.65 GHz Processor","4 GB RAM, 128 GB inbuilt",3110 mAh Battery,"6.1 inches, 828 x 1792 px Display with Large N...",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v13
149,Apple iPhone 12 Mini,"₹40,999",74.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 64 GB inbuilt","5.4 inches, 1080 x 2340 px Display",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14,No FM Radio


In [30]:
get_apple_phones(df).shape

(47, 11)

In [31]:
def get_ipods(df: pd.DataFrame):
  return (
      df
      .loc[(
          df["model"]
          .str.contains("iPod", flags=re.IGNORECASE)
      ), :]
  )

get_ipods(df)

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
754,Apple iPod Touch (7th Gen),"₹18,900",NaN,Wi-Fi,32 GB inbuilt,"4 inches, 640 x 1136 px Display",8 MP Rear & 1.2 MP Front Camera,iOS v12,No FM Radio,Bluetooth,Browser


In [32]:
~df.sample(10)["display"].str.contains("inches", flags=re.IGNORECASE)

840    False
963    False
430    False
330    False
473    False
920    False
990    False
303    False
951    False
362    False
Name: display, dtype: bool

In [33]:
def get_apple_shifted_phones(data: pd.DataFrame):
    apple_data = get_apple_phones(data)

    return (
        apple_data
        .loc[
            (~apple_data["display"].str.contains("inches", flags=re.IGNORECASE)) | \
            (~apple_data["camera"].str.contains("MP", flags=re.IGNORECASE)) | \
            (~apple_data["card"].str.contains("Card", flags=re.IGNORECASE))
        ]
    )


get_apple_shifted_phones(df)

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
111,Apple iPhone 12,"₹51,999",74.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 64 GB inbuilt","6.1 inches, 1170 x 2532 px Display with Large ...",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14,No FM Radio
149,Apple iPhone 12 Mini,"₹40,999",74.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 64 GB inbuilt","5.4 inches, 1080 x 2340 px Display",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14,No FM Radio
307,Apple iPhone 12 (128GB),"₹55,999",75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 128 GB inbuilt","6.1 inches, 1170 x 2532 px Display with Large ...",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14,No FM Radio
363,Apple iPhone 12 Mini (128GB),"₹45,999",75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 128 GB inbuilt","5.4 inches, 1080 x 2340 px Display",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14,No FM Radio
439,Apple iPhone SE 3 2022,"₹43,900",NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A15, Hexa Core, 3.22 GHz Processor",64 GB inbuilt,"4.7 inches, 750 x 1334 px Display",12 MP Rear & 7 MP Front Camera,Memory Card Not Supported,iOS v15,No FM Radio
448,Apple iPhone 15 Pro,"₹1,30,990",75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC",Bionic A16,"8 GB RAM, 128 GB inbuilt","6.06 inches, 1170 x 2532 px, 120 Hz Display wi...",50 MP + 12 MP + 12 MP Triple Rear & 12 MP Fron...,Memory Card Not Supported,iOS v15,No FM Radio
607,Apple iPhone 7s,"₹52,990",NaN,"Single Sim, 3G, 4G, VoLTE, Wi-Fi","Fusion APL1024, Quad Core, 2.37 GHz Processor","3 GB RAM, 32 GB inbuilt",2230 mAh Battery,"4.7 inches, 750 x 1334 px Display",13 MP Rear & 7 MP Front Camera,iOS v10,No FM Radio
628,Apple iPhone 12 Pro (512GB),"₹1,39,900",80.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","6 GB RAM, 512 GB inbuilt","6.1 inches, 1170 x 2532 px Display with Large ...",12 MP + 12 MP + 12 MP Triple Rear & 12 MP Fron...,Memory Card Not Supported,iOS v14.0,No FM Radio
656,Apple iPhone SE 2020,"₹39,900",63.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi, NFC","A13 Bionic, Hexa Core, 2.65 GHz Processor","3 GB RAM, 64 GB inbuilt",1821 mAh Battery with 18W Fast Charging,"4.7 inches, 750 x 1334 px Display",12 MP Rear & 7 MP Front Camera,iOS v13,No FM Radio
754,Apple iPod Touch (7th Gen),"₹18,900",NaN,Wi-Fi,32 GB inbuilt,"4 inches, 640 x 1136 px Display",8 MP Rear & 1.2 MP Front Camera,iOS v12,No FM Radio,Bluetooth,Browser


In [34]:
# remove feature phones from data

def remove_feature_phones(data: pd.DataFrame):

    temp_df = (
        data
        .drop(index=extract_feature_phones(data).index.to_list())
        .drop(index=get_ipods(data).index.to_list())
    )

    return temp_df

In [35]:
df_without_feature_phones = remove_feature_phones(df)

df_without_feature_phones.head()

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
0,OnePlus 11 5G,"₹54,999",89.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor","12 GB RAM, 256 GB inbuilt",5000 mAh Battery with 100W Fast Charging,"6.7 inches, 1440 x 3216 px, 120 Hz Display wit...",50 MP + 48 MP + 32 MP Triple Rear & 16 MP Fron...,Memory Card Not Supported,Android v13
1,OnePlus Nord CE 2 Lite 5G,"₹19,989",81.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 33W Fast Charging,"6.59 inches, 1080 x 2412 px, 120 Hz Display wi...",64 MP + 2 MP + 2 MP Triple Rear & 16 MP Front ...,"Memory Card (Hybrid), upto 1 TB",Android v12
2,Samsung Galaxy A14 5G,"₹16,499",75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Exynos 1330, Octa Core, 2.4 GHz Processor","4 GB RAM, 64 GB inbuilt",5000 mAh Battery with 15W Fast Charging,"6.6 inches, 1080 x 2408 px, 90 Hz Display with...",50 MP + 2 MP + 2 MP Triple Rear & 13 MP Front ...,"Memory Card Supported, upto 1 TB",Android v13
3,Motorola Moto G62 5G,"₹14,999",81.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with Fast Charging,"6.55 inches, 1080 x 2400 px, 120 Hz Display wi...",50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,"Memory Card (Hybrid), upto 1 TB",Android v12
4,Realme 10 Pro Plus,"₹24,999",82.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Dimensity 1080, Octa Core, 2.6 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 67W Fast Charging,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit...",108 MP + 8 MP + 2 MP Triple Rear & 16 MP Front...,Memory Card Not Supported,Android v13


In [36]:
df_without_feature_phones.shape

(976, 11)

In [37]:
# duplicates

df_without_feature_phones.duplicated().sum()

0

In [38]:
# missing values

df_without_feature_phones.isna().sum()

df_without_feature_phones.loc[df_without_feature_phones["os"].isna()]

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
645,Nokia 2760 Flip,"₹5,490",NaN,"Dual Sim, 3G, 4G, Wi-Fi",1450 mAh Battery,"3.6 inches, 240 x 320 px Display",5 MP Rear & 5 MP Front Camera,"Memory Card Supported, upto 32 GB",Kaios v3.0,Bluetooth,NaN
857,LG Folder 2,"₹11,999",NaN,"Single Sim, 3G, 4G, Wi-Fi","1 GB RAM, 8 GB inbuilt",1470 mAh Battery,"2.8 inches, 240 x 320 px Display",2 MP Rear Camera,Memory Card Supported,Bluetooth,NaN


In [39]:
def get_shifted_phones(data: pd.DataFrame):

    return (
        data
        .loc[
            (~data["display"].str.contains("inches", flags=re.IGNORECASE)) | \
            (~data["camera"].str.contains("MP", flags=re.IGNORECASE)) | \
            (~data["card"].str.contains("Card", flags=re.IGNORECASE))
        ]
    )

In [40]:
get_shifted_phones(df_without_feature_phones)

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
8,Nothing Phone 1,"₹26,749",85.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 778G Plus, Octa Core, 2.5 GHz Proce...","8 GB RAM, 128 GB inbuilt",4500 mAh Battery with 33W Fast Charging,"6.55 inches, 1080 x 2400 px, 120 Hz Display wi...",50 MP + 50 MP Dual Rear & 16 MP Front Camera,Android v12,No FM Radio
9,OnePlus Nord 2T 5G,"₹28,999",84.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Dimensity 1300, Octa Core, 3 GHz Processor","8 GB RAM, 128 GB inbuilt",4500 mAh Battery with 80W Fast Charging,"6.43 inches, 1080 x 2400 px, 90 Hz Display wit...",50 MP + 8 MP + 2 MP Triple Rear & 32 MP Front ...,Android v12,No FM Radio
12,Xiaomi Redmi Note 12 Pro 5G,"₹24,762",79.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster","Dimensity 1080, Octa Core, 2.6 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 67W Fast Charging,"6.67 inches, 1080 x 2400 px, 120 Hz Display wi...",50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,Android v12,No FM Radio
17,OPPO Reno 9 Pro Plus,"₹45,999",86.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8+ Gen1, Octa Core, 3.2 GHz Processor","16 GB RAM, 256 GB inbuilt",4700 mAh Battery with 80W Fast Charging,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit...",50 MP + 8 MP + 2 MP Triple Rear & 32 MP Front ...,Android v13,No FM Radio
18,OnePlus 10R 5G,"₹32,999",86.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Dimensity 8100 Max, Octa Core, 2.85 GHz Processor","8 GB RAM, 128 GB inbuilt",5000 mAh Battery with 80W Fast Charging,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit...",50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,Android v12,Bluetooth
...,...,...,...,...,...,...,...,...,...,...,...
1011,Oppo Find X6,"₹69,990",89.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor","8 GB RAM, 256 GB inbuilt",4700 mAh Battery with 120W Fast Charging,"6.73 inches, 1080 x 2400 px, 120 Hz Display wi...",50 MP + 48 MP + 32 MP Triple Rear & 32 MP Fron...,Android v12,No FM Radio
1012,itel A23s,"₹4,787",NaN,"Dual Sim, 3G, 4G, Wi-Fi","Spreadtrum SC9832E, Quad Core, 1.4 GHz Processor","2 GB RAM, 32 GB inbuilt",3020 mAh Battery,"5 inches, 854 x 480 px Display",2 MP Rear Camera,Android v11,No FM Radio
1013,Google Pixel 8 Pro,"₹70,990",80.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Google Tensor 3, Octa Core Processor","12 GB RAM, 256 GB inbuilt",5000 mAh Battery with 67W Fast Charging,"6.73 inches, 1440 x 3120 px, 120 Hz Display wi...",50 MP + 50 MP + 50 MP Triple Rear & 12 MP Fron...,Android v13,No FM Radio
1014,Vivo X Fold 2,"₹1,19,990",NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor","12 GB RAM, 256 GB inbuilt",4800 mAh Battery with 66W Fast Charging,"8.03 inches, 1916 x 2160 px, 120 Hz Display",Foldable Display,50 MP + 50 MP + 10 MP Triple Rear & 32 MP Fron...,Android v12


In [41]:
def get_left_shifted_phones(data: pd.DataFrame):

    return (
        data
        .loc[
            ~((data["os"].str.contains("Android", flags=re.IGNORECASE)) | \
             (data["os"].str.contains("iOS", flags=re.IGNORECASE)))
        ]
    )

get_left_shifted_phones(df_without_feature_phones)

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
8,Nothing Phone 1,"₹26,749",85.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 778G Plus, Octa Core, 2.5 GHz Proce...","8 GB RAM, 128 GB inbuilt",4500 mAh Battery with 33W Fast Charging,"6.55 inches, 1080 x 2400 px, 120 Hz Display wi...",50 MP + 50 MP Dual Rear & 16 MP Front Camera,Android v12,No FM Radio
9,OnePlus Nord 2T 5G,"₹28,999",84.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Dimensity 1300, Octa Core, 3 GHz Processor","8 GB RAM, 128 GB inbuilt",4500 mAh Battery with 80W Fast Charging,"6.43 inches, 1080 x 2400 px, 90 Hz Display wit...",50 MP + 8 MP + 2 MP Triple Rear & 32 MP Front ...,Android v12,No FM Radio
12,Xiaomi Redmi Note 12 Pro 5G,"₹24,762",79.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster","Dimensity 1080, Octa Core, 2.6 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 67W Fast Charging,"6.67 inches, 1080 x 2400 px, 120 Hz Display wi...",50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,Android v12,No FM Radio
17,OPPO Reno 9 Pro Plus,"₹45,999",86.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8+ Gen1, Octa Core, 3.2 GHz Processor","16 GB RAM, 256 GB inbuilt",4700 mAh Battery with 80W Fast Charging,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit...",50 MP + 8 MP + 2 MP Triple Rear & 32 MP Front ...,Android v13,No FM Radio
18,OnePlus 10R 5G,"₹32,999",86.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Dimensity 8100 Max, Octa Core, 2.85 GHz Processor","8 GB RAM, 128 GB inbuilt",5000 mAh Battery with 80W Fast Charging,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit...",50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,Android v12,Bluetooth
...,...,...,...,...,...,...,...,...,...,...,...
1009,Xiaomi Civi 3,"₹32,990",86.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...","Dimensity 8200, Octa Core, 3.1 GHz Processor","8 GB RAM, 256 GB inbuilt",5000 mAh Battery with 80W Fast Charging,"6.7 inches, 1080 x 2400 px, 120 Hz Display wit...",64 MP + 20 MP + 2 MP Triple Rear & 32 MP + 32 ...,Android v13,No FM Radio
1011,Oppo Find X6,"₹69,990",89.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor","8 GB RAM, 256 GB inbuilt",4700 mAh Battery with 120W Fast Charging,"6.73 inches, 1080 x 2400 px, 120 Hz Display wi...",50 MP + 48 MP + 32 MP Triple Rear & 32 MP Fron...,Android v12,No FM Radio
1012,itel A23s,"₹4,787",NaN,"Dual Sim, 3G, 4G, Wi-Fi","Spreadtrum SC9832E, Quad Core, 1.4 GHz Processor","2 GB RAM, 32 GB inbuilt",3020 mAh Battery,"5 inches, 854 x 480 px Display",2 MP Rear Camera,Android v11,No FM Radio
1013,Google Pixel 8 Pro,"₹70,990",80.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Google Tensor 3, Octa Core Processor","12 GB RAM, 256 GB inbuilt",5000 mAh Battery with 67W Fast Charging,"6.73 inches, 1440 x 3120 px, 120 Hz Display wi...",50 MP + 50 MP + 50 MP Triple Rear & 12 MP Fron...,Android v13,No FM Radio


**Subset of Data**:

Things to remove:
1. feature phones & ipods --> completed
2. shifted phones
3. incomplete data

In [43]:
indexes_of_shifted_phones = get_shifted_phones(df_without_feature_phones).index.to_list()

len(indexes_of_shifted_phones)

251

In [44]:
indexes_of_incomplete_data = incomplete_data_phones(df_without_feature_phones).index.to_list()

len(indexes_of_incomplete_data)

390

In [45]:
indexes_to_drop = set(indexes_of_shifted_phones)

In [46]:
len(indexes_to_drop)

251

In [47]:
final_df = df_without_feature_phones.drop(index=indexes_to_drop)

final_df.head()

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
0,OnePlus 11 5G,"₹54,999",89.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor","12 GB RAM, 256 GB inbuilt",5000 mAh Battery with 100W Fast Charging,"6.7 inches, 1440 x 3216 px, 120 Hz Display wit...",50 MP + 48 MP + 32 MP Triple Rear & 16 MP Fron...,Memory Card Not Supported,Android v13
1,OnePlus Nord CE 2 Lite 5G,"₹19,989",81.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 33W Fast Charging,"6.59 inches, 1080 x 2412 px, 120 Hz Display wi...",64 MP + 2 MP + 2 MP Triple Rear & 16 MP Front ...,"Memory Card (Hybrid), upto 1 TB",Android v12
2,Samsung Galaxy A14 5G,"₹16,499",75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Exynos 1330, Octa Core, 2.4 GHz Processor","4 GB RAM, 64 GB inbuilt",5000 mAh Battery with 15W Fast Charging,"6.6 inches, 1080 x 2408 px, 90 Hz Display with...",50 MP + 2 MP + 2 MP Triple Rear & 13 MP Front ...,"Memory Card Supported, upto 1 TB",Android v13
3,Motorola Moto G62 5G,"₹14,999",81.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with Fast Charging,"6.55 inches, 1080 x 2400 px, 120 Hz Display wi...",50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,"Memory Card (Hybrid), upto 1 TB",Android v12
4,Realme 10 Pro Plus,"₹24,999",82.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Dimensity 1080, Octa Core, 2.6 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 67W Fast Charging,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit...",108 MP + 8 MP + 2 MP Triple Rear & 16 MP Front...,Memory Card Not Supported,Android v13


In [48]:
final_df.shape

(725, 11)

In [49]:
725 / 1020

0.7107843137254902

## Column Wise Cleaning

### Model

In [52]:
df.columns

Index(['model', 'price', 'rating', 'sim', 'processor', 'ram', 'battery',
       'display', 'camera', 'card', 'os'],
      dtype='object')

In [53]:
final_df["model"]

0                                 OnePlus 11 5G
1                     OnePlus Nord CE 2 Lite 5G
2                         Samsung Galaxy A14 5G
3                          Motorola Moto G62 5G
4                            Realme 10 Pro Plus
                         ...                   
1010    Realme Narzo 50i Prime (4GB RAM + 64GB)
1016                                Honor X8 5G
1017            POCO X4 GT 5G (8GB RAM + 256GB)
1018                       Motorola Moto G91 5G
1019                     Samsung Galaxy M52s 5G
Name: model, Length: 725, dtype: object

In [54]:
# separated out brand and model_name from model

(
    final_df["model"]
    .str.split(" ", n=1, expand=True)
    .set_axis(labels=["brand", "model_name"], axis=1)
    .assign(
        model_name=lambda df_: df_["model_name"].str.replace("\(.+\)", "", regex=True)
        )
)

<>:8: SyntaxWarning: invalid escape sequence '\('
<>:8: SyntaxWarning: invalid escape sequence '\('
C:\Users\dipak\AppData\Local\Temp\ipykernel_18756\1236067222.py:8: SyntaxWarning: invalid escape sequence '\('
  model_name=lambda df_: df_["model_name"].str.replace("\(.+\)", "", regex=True)


,brand,model_name
0,OnePlus,11 5G
1,OnePlus,Nord CE 2 Lite 5G
2,Samsung,Galaxy A14 5G
3,Motorola,Moto G62 5G
4,Realme,10 Pro Plus
...,...,...
1010,Realme,Narzo 50i Prime
1016,Honor,X8 5G
1017,POCO,X4 GT 5G
1018,Motorola,Moto G91 5G


In [55]:
(
    final_df["model"]
    .str.split(" ", n=1, expand=True)
    .set_axis(labels=["brand", "model_name"], axis=1)
    .assign(
        model_name=lambda df_: df_["model_name"].str.replace("\(.+\)", "", regex=True)
        )
    .loc[:, "brand"]
    .str.title()
    .unique()
)

<>:6: SyntaxWarning: invalid escape sequence '\('
<>:6: SyntaxWarning: invalid escape sequence '\('
C:\Users\dipak\AppData\Local\Temp\ipykernel_18756\2478932270.py:6: SyntaxWarning: invalid escape sequence '\('
  model_name=lambda df_: df_["model_name"].str.replace("\(.+\)", "", regex=True)


array(['Oneplus', 'Samsung', 'Motorola', 'Realme', 'Apple', 'Xiaomi',
       'Oppo', 'Vivo', 'Poco', 'Jio', 'Iqoo', 'Gionee', 'Tesla', 'Google',
       'Infinix', 'Cola', 'Tecno', 'Ikall', 'Nokia', 'Lava', 'Redmi',
       'Asus', 'Itel', 'Sony', 'Oukitel', 'Vertu', 'Blu', 'Huawei',
       'Micromax', 'Lg', 'Honor', 'Leitz', 'Doogee', 'Zte', 'Tcl',
       'Sharp', 'Blackview'], dtype=object)

In [56]:
(
    final_df["model"]
    .str.split(" ", n=1, expand=True)
    .set_axis(labels=["brand", "model_name"], axis=1)
    .assign(
        model_name=lambda df_: df_["model_name"].str.replace("\(.+\)", "", regex=True)
        )
    .loc[:,"model_name"]
    .str.contains("[\(\)]", regex=True)
    .sum()
)

<>:6: SyntaxWarning: invalid escape sequence '\('
<>:9: SyntaxWarning: invalid escape sequence '\('
<>:6: SyntaxWarning: invalid escape sequence '\('
<>:9: SyntaxWarning: invalid escape sequence '\('
C:\Users\dipak\AppData\Local\Temp\ipykernel_18756\3135688989.py:6: SyntaxWarning: invalid escape sequence '\('
  model_name=lambda df_: df_["model_name"].str.replace("\(.+\)", "", regex=True)
C:\Users\dipak\AppData\Local\Temp\ipykernel_18756\3135688989.py:9: SyntaxWarning: invalid escape sequence '\('
  .str.contains("[\(\)]", regex=True)


0

In [57]:
def clean_model(ser:pd.Series) -> pd.DataFrame:
    return (
            ser
            .str.split(" ", n=1, expand=True)
            .set_axis(labels=["brand", "model_name"], axis=1)
            .assign(
                    model_name=lambda df_: (df_["model_name"]
                                            .str.replace("\(.+\)", "", regex=True)
                                            .str.strip()),
                    brand=lambda df_: df_["brand"].str.title().str.strip().str.title()
                    )
    )

<>:8: SyntaxWarning: invalid escape sequence '\('
<>:8: SyntaxWarning: invalid escape sequence '\('
C:\Users\dipak\AppData\Local\Temp\ipykernel_18756\2680617267.py:8: SyntaxWarning: invalid escape sequence '\('
  .str.replace("\(.+\)", "", regex=True)


In [58]:
clean_model(final_df["model"])

,brand,model_name
0,Oneplus,11 5G
1,Oneplus,Nord CE 2 Lite 5G
2,Samsung,Galaxy A14 5G
3,Motorola,Moto G62 5G
4,Realme,10 Pro Plus
...,...,...
1010,Realme,Narzo 50i Prime
1016,Honor,X8 5G
1017,Poco,X4 GT 5G
1018,Motorola,Moto G91 5G


In [59]:
clean_model(final_df["model"]).loc[:,"brand"].unique()

array(['Oneplus', 'Samsung', 'Motorola', 'Realme', 'Apple', 'Xiaomi',
       'Oppo', 'Vivo', 'Poco', 'Jio', 'Iqoo', 'Gionee', 'Tesla', 'Google',
       'Infinix', 'Cola', 'Tecno', 'Ikall', 'Nokia', 'Lava', 'Redmi',
       'Asus', 'Itel', 'Sony', 'Oukitel', 'Vertu', 'Blu', 'Huawei',
       'Micromax', 'Lg', 'Honor', 'Leitz', 'Doogee', 'Zte', 'Tcl',
       'Sharp', 'Blackview'], dtype=object)

### Price

In [61]:
final_df["price"]

0       ₹54,999
1       ₹19,989
2       ₹16,499
3       ₹14,999
4       ₹24,999
         ...   
1010     ₹8,720
1016    ₹14,990
1017    ₹28,990
1018    ₹19,990
1019    ₹24,990
Name: price, Length: 725, dtype: object

In [62]:
def clean_price(ser):

    return (
        ser
        .str.replace("₹", "")
        .str.replace(",", "")
        .astype(np.float64)
    )

In [63]:
type(clean_price(final_df["price"].sample(10)))

pandas.core.series.Series

### Rating

In [65]:
final_df["rating"]

0       89.0
1       81.0
2       75.0
3       81.0
4       82.0
        ... 
1010    64.0
1016    75.0
1017    85.0
1018    80.0
1019    74.0
Name: rating, Length: 725, dtype: float64

In [66]:
# missing values in rating

final_df["rating"].isna().sum()

49

In [67]:
final_df[final_df["rating"].isna()]

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
14,Samsung Galaxy S23 Ultra 5G,"₹1,14,990",NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC","Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor","8 GB RAM, 256 GB inbuilt",5000 mAh Battery with 45W Fast Charging,"6.8 inches, 1440 x 3088 px, 120 Hz Display wit...",200 MP Quad Rear & 12 MP Front Camera,Memory Card Not Supported,Android v13
49,Samsung Galaxy A74 5G,"₹42,999",NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 7 Gen1, Octa Core, 2.36 GHz Processor","8 GB RAM, 128 GB inbuilt",5000 mAh Battery with 33W Fast Charging,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit...",108 MP Quad Rear & 32 MP Front Camera,"Memory Card (Hybrid), upto 1 TB",Android v12
75,Gionee G13 Pro,"₹6,190",NaN,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Tiger T310, Quad Core, 2 GHz Processor","4 GB RAM, 32 GB inbuilt",3500 mAh Battery,"6.26 inches, 720 x 1600 px Display with Large ...",13 MP Dual Rear & 5 MP Front Camera,Memory Card Supported,HarmonyOS v2
89,Samsung Galaxy A73 5G,"₹41,999",NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 778G, Octa Core, 2.4 GHz Processor","8 GB RAM, 128 GB inbuilt",5000 mAh Battery with 25W Fast Charging,"6.7 inches, 1080 x 2400 px, 120 Hz Display wit...",108 MP Quad Rear & 32 MP Front Camera,"Memory Card (Hybrid), upto 1 TB",Android v12
105,Xiaomi 13 Pro 5G,"₹58,990",NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...","Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor","8 GB RAM, 128 GB inbuilt",4820 mAh Battery with 120W Fast Charging,"6.73 inches, 1440 x 3200 px, 120 Hz Display wi...",50.3 MP + 50 MP + 50 MP Triple Rear & 32 MP Fr...,Memory Card Not Supported,Android v13
114,Xiaomi Redmi A1,"₹6,171",NaN,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Helio A22, Quad Core, 2 GHz Processor","2 GB RAM, 32 GB inbuilt",5000 mAh Battery with 10W Fast Charging,"6.52 inches, 720 x 1600 px Display with Water ...",8 MP Dual Rear & 5 MP Front Camera,"Memory Card Supported, upto 512 GB",Android v12
143,Jio Phone 3,"₹4,499",NaN,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Quad Core, 1.4 GHz Processor","2 GB RAM, 64 GB inbuilt",2800 mAh Battery,"5 inches, 720 x 1280 px Display",5 MP Rear & 2 MP Front Camera,"Memory Card Supported, upto 128 GB",Android v8.1 (Oreo)
156,Realme C30,"₹5,299",NaN,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Tiger T612, Octa Core, 1.82 GHz Processor","2 GB RAM, 32 GB inbuilt",5000 mAh Battery,"6.5 inches, 720 x 1600 px Display with Water D...",8 MP Rear & 5 MP Front Camera,"Memory Card Supported, upto 1 TB",Android v11
193,Nokia N73 5G,"₹46,999",NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 8 Gen1, Octa Core, 3 GHz Processor","12 GB RAM, 256 GB inbuilt",6000 mAh Battery with 65W Fast Charging,"6.9 inches, 1400 x 3200 px, 120 Hz Display wit...",200 MP Penta Rear & 32 MP Front Camera,Memory Card Supported,Android v12.1
225,Vivo Y02,"₹8,999",NaN,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Helio, Octa Core Processor","3 GB RAM, 32 GB inbuilt",5000 mAh Battery with 10W Fast Charging,"6.51 inches, 720 x 1600 px Display with Water ...",8 MP Rear & 5 MP Front Camera,"Memory Card Supported, upto 1 TB",Android v12


### Sim

In [69]:
final_df["sim"]

0                 Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC
1                      Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi
2                      Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi
3                      Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi
4                      Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi
                              ...                        
1010                       Dual Sim, 3G, 4G, VoLTE, Wi-Fi
1016                   Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi
1017    Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...
1018              Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC
1019                   Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi
Name: sim, Length: 725, dtype: object

In [70]:
final_df.shape

(725, 11)

In [71]:
# 4G phones in the data

(
    final_df["sim"]
    .str.contains("4G")
    .sum()
)

724

In [72]:
# phone that is non 4G

(
    final_df["sim"]
    .str.contains("4G")
    .loc[lambda x: x == False]
)

735    False
Name: sim, dtype: bool

In [73]:
final_df.loc[735]

model                                           Tecno Pop 5 Go
price                                                   ₹6,999
rating                                                     NaN
sim                                        Dual Sim, 3G, Wi-Fi
processor     Spreadtrum  SC7731, Quad Core, 1.3 GHz Processor
ram                                    1 GB RAM, 16 GB inbuilt
battery                                       4000 mAh Battery
display      6.1 inches, 720 x 1520 px Display with Water D...
camera             5 MP + 0.3 MP Dual Rear & 5 MP Front Camera
card                                     Memory Card Supported
os                                                 Android v11
Name: 735, dtype: object

In [74]:
# number of sims

(
final_df["sim"]
.str.extract("(\w+) Sim",flags=re.IGNORECASE)
.value_counts()
)

<>:5: SyntaxWarning: invalid escape sequence '\w'
<>:5: SyntaxWarning: invalid escape sequence '\w'
C:\Users\dipak\AppData\Local\Temp\ipykernel_18756\615634867.py:5: SyntaxWarning: invalid escape sequence '\w'
  .str.extract("(\w+) Sim",flags=re.IGNORECASE)


0     
Dual      718
Single      7
Name: count, dtype: int64

In [75]:
(
    final_df["sim"]
    .str.extract("(\w+) Sim",flags=re.IGNORECASE)
    .squeeze()
    .loc[lambda ser: ser == "Single"]
    .index
)

<>:3: SyntaxWarning: invalid escape sequence '\w'
<>:3: SyntaxWarning: invalid escape sequence '\w'
C:\Users\dipak\AppData\Local\Temp\ipykernel_18756\3403282238.py:3: SyntaxWarning: invalid escape sequence '\w'
  .str.extract("(\w+) Sim",flags=re.IGNORECASE)


Index([174, 429, 431, 457, 506, 870, 899], dtype='int64')

In [76]:
final_df.loc[[174, 429, 431, 457, 506, 870, 899]]

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
174,Apple iPhone 9,"₹29,990",61.0,"Single Sim, 3G, 4G, VoLTE, Wi-Fi, NFC","A13 Bionic, Hexa Core, 2.65 GHz Processor","3 GB RAM, 64 GB inbuilt",2050 mAh Battery with Fast Charging,"4.7 inches, 750 x 1334 px Display with Large N...",12 MP Rear & 7 MP Front Camera,Memory Card Not Supported,iOS v13.0
429,Nokia X50 5G,"₹34,999",76.0,"Single Sim, 3G, 4G, 5G, VoLTE, Wi-Fi",Snapdragon 775,"6 GB RAM, 64 GB inbuilt",6000 mAh Battery with 33W Fast Charging,"6.81 inches, 1080 x 2400 px Display with Punch...",108 MP Quad Rear & 32 MP Front Camera,Memory Card Supported,Android v11
431,Vertu Signature Touch,"₹6,50,000",62.0,"Single Sim, 3G, 4G, Wi-Fi, NFC","Snapdragon 801, Octa Core, 1.5 GHz Processor","2 GB RAM, 64 GB inbuilt",2275 mAh Battery,"4.7 inches, 1080 x 1920 px Display",13 MP Rear & 2.1 MP Front Camera,Memory Card Not Supported,Android v4.4.2 (KitKat)
457,Sony Xperia L5 5G,"₹15,990",73.0,"Single Sim, 3G, 4G, VoLTE, Wi-Fi, NFC","Helio P35, Octa Core, 2.3 GHz Processor","4 GB RAM, 64 GB inbuilt",4000 mAh Battery with Fast Charging,"6.2 inches, 720 x 1680 px Display",13 MP + 5 MP + 2 MP Triple Rear & 8 MP Front C...,"Memory Card (Hybrid), upto 512 GB",Android v12
506,Google Pixel 2 XL,"₹15,990",69.0,"Single Sim, 3G, 4G, VoLTE, Wi-Fi, NFC","Snapdragon 835, Octa Core, 2.35 GHz Processor","4 GB RAM, 128 GB inbuilt",3520 mAh Battery with Fast Charging,"6 inches, 1440 x 2880 px Display",12.2 MP Rear & 8 MP Front Camera,Memory Card Not Supported,Android v8.0 (Oreo)
870,Samsung Galaxy S10 5G,"₹78,990",86.0,"Single Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Exynos 9820, Octa Core, 2.7 GHz Processor","8 GB RAM, 256 GB inbuilt",4500 mAh Battery with 25W Fast Charging,"6.7 inches, 1440 x 3040 px Display",16 MP + 12 MP + 12 MP Triple Rear & 10 MP Fron...,"Memory Card (Hybrid), upto 512 GB",Android v9.0 (Pie)
899,Sony Xperia 10 III Lite 5G,"₹30,990",78.0,"Single Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 690 , Octa Core, 2 GHz Processor","6 GB RAM, 64 GB inbuilt",4500 mAh Battery with 30W Fast Charging,"6 inches, 1080 x 2520 px Display",12 MP + 8 MP + 8 MP Triple Rear & 8 MP Front C...,Memory Card Supported,Android v11


In [77]:
def clean_sim(ser: pd.Series) -> pd.DataFrame:
    result_df = pd.DataFrame()

    return (
        result_df
        .assign(
            number_of_sim=ser.str.extract("(\w+) Sim",flags=re.IGNORECASE),
            has_5G=ser.str.contains("5G", flags=re.IGNORECASE),
            has_NFC=ser.str.contains("NFC", flags=re.IGNORECASE),
            has_IR_Blaster=ser.str.contains("IR Blaster", flags=re.IGNORECASE)
        )
        .assign(
            number_of_sim=lambda df_: np.where(df_["number_of_sim"] == "Dual", 2, 1)
        )
    )

<>:7: SyntaxWarning: invalid escape sequence '\w'
<>:7: SyntaxWarning: invalid escape sequence '\w'
C:\Users\dipak\AppData\Local\Temp\ipykernel_18756\477182850.py:7: SyntaxWarning: invalid escape sequence '\w'
  number_of_sim=ser.str.extract("(\w+) Sim",flags=re.IGNORECASE),


In [78]:
clean_sim(final_df["sim"]).shape

(725, 4)

In [79]:
clean_sim(final_df["sim"]).isna().sum()

number_of_sim     0
has_5G            0
has_NFC           0
has_IR_Blaster    0
dtype: int64

### Processor

In [81]:
(
    final_df["processor"]
)

0       Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor
1          Snapdragon 695, Octa Core, 2.2 GHz Processor
2             Exynos 1330, Octa Core, 2.4 GHz Processor
3         Snapdragon  695, Octa Core, 2.2 GHz Processor
4          Dimensity 1080, Octa Core, 2.6 GHz Processor
                             ...                       
1010          Tiger T612, Octa Core, 1.82 GHz Processor
1016     Snapdragon  480+, Octa Core, 2.2 GHz Processor
1017      Dimensity 8100, Octa Core, 2.85 GHz Processor
1018       Snapdragon 695, Octa Core, 2.2 GHz Processor
1019                                Octa Core Processor
Name: processor, Length: 725, dtype: object

In [82]:
def clean_processor(ser: pd.Series):
    temp_df = pd.DataFrame()

    return (
        temp_df
        .assign(
            cpu_speed=ser.str.extract("(\d+\.?\d*).?GHz").astype(np.float64),
            number_of_cores=ser.str.extract("(\w+)\sCore"),
            processor_name=np.where(ser.str.count(",").eq(2), ser.str.split(",", n=1).str.get(0), np.nan)
        )
        .assign(
            number_of_cores= lambda df_: np.select(condlist=[df_["number_of_cores"].str.contains("Octa", case=False, na=False),
                                                            df_["number_of_cores"].str.contains("Hexa", case=False, na=False),
                                                            df_["number_of_cores"].str.contains("Quad", case=False, na=False),
                                                            df_["number_of_cores"].str.contains("Dual", case=False, na=False)],
                                                            choicelist=[8, 6, 4, 2],
                                                            default=1),
            processor_name=lambda df_: df_["processor_name"].str.split(" ", n=1).str.get(0)
        )
    )


<>:7: SyntaxWarning: invalid escape sequence '\d'
<>:8: SyntaxWarning: invalid escape sequence '\w'
<>:7: SyntaxWarning: invalid escape sequence '\d'
<>:8: SyntaxWarning: invalid escape sequence '\w'
C:\Users\dipak\AppData\Local\Temp\ipykernel_18756\3987612111.py:7: SyntaxWarning: invalid escape sequence '\d'
  cpu_speed=ser.str.extract("(\d+\.?\d*).?GHz").astype(np.float64),
C:\Users\dipak\AppData\Local\Temp\ipykernel_18756\3987612111.py:8: SyntaxWarning: invalid escape sequence '\w'
  number_of_cores=ser.str.extract("(\w+)\sCore"),


In [83]:
clean_processor(final_df["processor"]).join(final_df["processor"])

,cpu_speed,number_of_cores,processor_name,processor
0,3.20,8,Snapdragon,"Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor"
1,2.20,8,Snapdragon,"Snapdragon 695, Octa Core, 2.2 GHz Processor"
2,2.40,8,Exynos,"Exynos 1330, Octa Core, 2.4 GHz Processor"
3,2.20,8,Snapdragon,"Snapdragon 695, Octa Core, 2.2 GHz Processor"
4,2.60,8,Dimensity,"Dimensity 1080, Octa Core, 2.6 GHz Processor"
...,...,...,...,...
1010,1.82,8,Tiger,"Tiger T612, Octa Core, 1.82 GHz Processor"
1016,2.20,8,Snapdragon,"Snapdragon 480+, Octa Core, 2.2 GHz Processor"
1017,2.85,8,Dimensity,"Dimensity 8100, Octa Core, 2.85 GHz Processor"
1018,2.20,8,Snapdragon,"Snapdragon 695, Octa Core, 2.2 GHz Processor"


In [84]:
clean_processor(final_df["processor"]).isna().sum()

cpu_speed          36
number_of_cores     0
processor_name     47
dtype: int64

In [85]:
clean_processor(final_df["processor"]).dtypes

cpu_speed          float64
number_of_cores      int32
processor_name      object
dtype: object

In [86]:
(
    clean_processor(final_df["processor"])
    .loc[clean_processor(final_df["processor"]).isna().any(axis=1),:]
    .join(final_df["processor"])
)

,cpu_speed,number_of_cores,processor_name,processor
27,NaN,6,NaN,"Bionic A16, Hexa Core Processor"
78,NaN,8,NaN,"Exynos 1280, Octa Core Processor"
100,NaN,6,NaN,"Bionic A16, Hexa Core Processor"
118,NaN,8,NaN,Octa Core Processor
122,NaN,8,NaN,"Google Tensor, Octa Core Processor"
129,NaN,8,NaN,"Google Tensor, Octa Core Processor"
143,1.4,4,NaN,"Quad Core, 1.4 GHz Processor"
152,NaN,1,NaN,Bionic A16
188,1.6,4,NaN,"Quad Core, 1.6 GHz Processor"
201,2.0,8,NaN,"Octa Core, 2 GHz Processor"


### Ram

In [88]:
final_df["ram"]

0       12 GB RAM, 256 GB inbuilt
1        6 GB RAM, 128 GB inbuilt
2         4 GB RAM, 64 GB inbuilt
3        6 GB RAM, 128 GB inbuilt
4        6 GB RAM, 128 GB inbuilt
                  ...            
1010      4 GB RAM, 64 GB inbuilt
1016     6 GB RAM, 128 GB inbuilt
1017     8 GB RAM, 256 GB inbuilt
1018     6 GB RAM, 128 GB inbuilt
1019     8 GB RAM, 128 GB inbuilt
Name: ram, Length: 725, dtype: object

In [89]:
(
    final_df["ram"]
    .str.findall("(\d+).?(?:GB|TB)")
    .to_frame()
    .rename(columns={"ram": "memory"})
    .assign(
        ram= lambda df_: df_["memory"].str.get(0),
        storage=lambda df_: df_["memory"].str.get(1)
    )
)

<>:3: SyntaxWarning: invalid escape sequence '\d'
<>:3: SyntaxWarning: invalid escape sequence '\d'
C:\Users\dipak\AppData\Local\Temp\ipykernel_18756\2266183211.py:3: SyntaxWarning: invalid escape sequence '\d'
  .str.findall("(\d+).?(?:GB|TB)")


,memory,ram,storage
0,"[12, 256]",12,256
1,"[6, 128]",6,128
2,"[4, 64]",4,64
3,"[6, 128]",6,128
4,"[6, 128]",6,128
...,...,...,...
1010,"[4, 64]",4,64
1016,"[6, 128]",6,128
1017,"[8, 256]",8,256
1018,"[6, 128]",6,128


In [90]:
final_df["ram"]

0       12 GB RAM, 256 GB inbuilt
1        6 GB RAM, 128 GB inbuilt
2         4 GB RAM, 64 GB inbuilt
3        6 GB RAM, 128 GB inbuilt
4        6 GB RAM, 128 GB inbuilt
                  ...            
1010      4 GB RAM, 64 GB inbuilt
1016     6 GB RAM, 128 GB inbuilt
1017     8 GB RAM, 256 GB inbuilt
1018     6 GB RAM, 128 GB inbuilt
1019     8 GB RAM, 128 GB inbuilt
Name: ram, Length: 725, dtype: object

In [91]:
def clean_ram(ser: pd.Series):
    temp_df = pd.DataFrame()
    return (
        temp_df
        .assign(
            memory= lambda df_: (ser.
                                 str.extract("(\d+).?(?:GB|TB)\sRAM", flags=re.IGNORECASE)
                                 .astype(np.float64)),
            storage=lambda df_: (ser
                                 .str.extract("(\d+).?(?:GB|TB)\sinbuilt", flags=re.IGNORECASE)
                                 .astype(np.float64))
                                 .replace(1.0, 1024.0)
        )

    )

<>:7: SyntaxWarning: invalid escape sequence '\d'
<>:10: SyntaxWarning: invalid escape sequence '\d'
<>:7: SyntaxWarning: invalid escape sequence '\d'
<>:10: SyntaxWarning: invalid escape sequence '\d'
C:\Users\dipak\AppData\Local\Temp\ipykernel_18756\840458078.py:7: SyntaxWarning: invalid escape sequence '\d'
  str.extract("(\d+).?(?:GB|TB)\sRAM", flags=re.IGNORECASE)
C:\Users\dipak\AppData\Local\Temp\ipykernel_18756\840458078.py:10: SyntaxWarning: invalid escape sequence '\d'
  .str.extract("(\d+).?(?:GB|TB)\sinbuilt", flags=re.IGNORECASE)


In [92]:
clean_ram(final_df["ram"])

,memory,storage
0,12.0,256.0
1,6.0,128.0
2,4.0,64.0
3,6.0,128.0
4,6.0,128.0
...,...,...
1010,4.0,64.0
1016,6.0,128.0
1017,8.0,256.0
1018,6.0,128.0


In [93]:
clean_ram(final_df["ram"]).isna().sum()

memory     1
storage    0
dtype: int64

In [94]:
clean_ram(final_df["ram"])["storage"].unique()

array([ 256.,  128.,   64.,   32.,  512., 1024.,   16.])

In [95]:
clean_ram(final_df["ram"])["memory"].unique()

array([12.,  6.,  4.,  8.,  3.,  2., 16., 18., nan,  1.])

In [96]:
final_df.loc[clean_ram(final_df["ram"])["storage"] == 1024]

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
290,Apple iPhone 14 Pro Max (1TB),"₹1,82,999",78.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A16, Hexa Core Processor","6 GB RAM, 1 TB inbuilt",4323 mAh Battery with Fast Charging,"6.7 inches, 1290 x 2796 px, 120 Hz Display",48 MP + 12 MP + 12 MP Triple Rear & 12 MP Fron...,Memory Card Not Supported,iOS v16
781,Apple iPhone 13 Pro Max (1TB),"₹1,79,900",86.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A15, Hexa Core, 3.22 GHz Processor","6 GB RAM, 1 TB inbuilt",4352 mAh Battery with Fast Charging,"6.7 inches, 1284 x 2778 px, 120 Hz Display wit...",12 MP + 12 MP + 12 MP Triple Rear & 12 MP Fron...,Memory Card Not Supported,iOS v15
814,Apple iPhone 14 Pro (1TB),"₹1,72,999",77.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A16, Hexa Core Processor","6 GB RAM, 1 TB inbuilt",3200 mAh Battery with Fast Charging,"6.1 inches, 1179 x 2556 px, 120 Hz Display",48 MP + 12 MP + 12 MP Triple Rear & 12 MP Fron...,Memory Card Not Supported,iOS v16
961,Apple iPhone 13 Pro (1TB),"₹1,47,900",84.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A15, Hexa Core, 3.22 GHz Processor","6 GB RAM, 1 TB inbuilt",3095 mAh Battery with Fast Charging,"6.1 inches, 1170 x 2532 px, 120 Hz Display wit...",12 MP + 12 MP + 12 MP Triple Rear & 12 MP Fron...,Memory Card Not Supported,iOS v15.0


In [97]:
final_df.loc[(
    clean_ram(final_df["ram"])
    .isna()
    .any(axis=1)
    )]

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
483,Huawei Mate 50 RS Porsche Design,"₹2,39,999",81.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi, NFC, IR Blaster","Snapdragon 8+ Gen1, Octa Core, 3.2 GHz Processor",512 GB inbuilt,4700 mAh Battery with 66W Fast Charging,"6.74 inches, 1212 x 2616 px, 120 Hz Display",50 MP + 48 MP + 13 MP Triple Rear & 13 MP Fron...,"Memory Card (Hybrid), upto 256 GB",Hongmeng OS v3.0


In [98]:
df.columns

Index(['model', 'price', 'rating', 'sim', 'processor', 'ram', 'battery',
       'display', 'camera', 'card', 'os'],
      dtype='object')

### Battery

In [100]:
def clean_battery(ser: pd.Series):
    temp_df = pd.DataFrame()

    return(
        temp_df
        .assign(
            capacity=ser.str.extract("(\d+).?mAh"),
            fast_charging=ser.str.contains("Fast Charging", flags=re.IGNORECASE),
            fast_charging_power=ser.str.extract("(\d+)\s?W")
        )

    )

<>:7: SyntaxWarning: invalid escape sequence '\d'
<>:9: SyntaxWarning: invalid escape sequence '\d'
<>:7: SyntaxWarning: invalid escape sequence '\d'
<>:9: SyntaxWarning: invalid escape sequence '\d'
C:\Users\dipak\AppData\Local\Temp\ipykernel_18756\3416880634.py:7: SyntaxWarning: invalid escape sequence '\d'
  capacity=ser.str.extract("(\d+).?mAh"),
C:\Users\dipak\AppData\Local\Temp\ipykernel_18756\3416880634.py:9: SyntaxWarning: invalid escape sequence '\d'
  fast_charging_power=ser.str.extract("(\d+)\s?W")


In [101]:
clean_battery(final_df["battery"]).join(final_df["battery"])

,capacity,fast_charging,fast_charging_power,battery
0,5000,True,100,5000 mAh Battery with 100W Fast Charging
1,5000,True,33,5000 mAh Battery with 33W Fast Charging
2,5000,True,15,5000 mAh Battery with 15W Fast Charging
3,5000,True,NaN,5000 mAh Battery with Fast Charging
4,5000,True,67,5000 mAh Battery with 67W Fast Charging
...,...,...,...,...
1010,5000,True,10,5000 mAh Battery with 10W Fast Charging
1016,5000,True,5,5000 mAh Battery with 22.5W Fast Charging
1017,5080,True,67,5080 mAh Battery with 67W Fast Charging
1018,5000,True,NaN,5000 mAh Battery with Fast Charging


In [102]:
clean_battery(final_df["battery"]).isna().sum()

capacity                 0
fast_charging            0
fast_charging_power    181
dtype: int64

In [103]:
final_df.loc[clean_battery(final_df["battery"])["fast_charging_power"].isna(),["battery"]].join(clean_battery(final_df["battery"]))

,battery,capacity,fast_charging,fast_charging_power
3,5000 mAh Battery with Fast Charging,5000,True,NaN
6,3279 mAh Battery with Fast Charging,3279,True,NaN
15,3240 mAh Battery with Fast Charging,3240,True,NaN
25,4500 mAh Battery with Fast Charging,4500,True,NaN
27,4323 mAh Battery with Fast Charging,4323,True,NaN
...,...,...,...,...
1003,4000 mAh Battery,4000,False,NaN
1005,5000 mAh Battery,5000,False,NaN
1006,8000 mAh Battery,8000,False,NaN
1018,5000 mAh Battery with Fast Charging,5000,True,NaN


In [104]:
(
    final_df["battery"]
)

0        5000 mAh Battery with 100W Fast Charging
1         5000 mAh Battery with 33W Fast Charging
2         5000 mAh Battery with 15W Fast Charging
3             5000 mAh Battery with Fast Charging
4         5000 mAh Battery with 67W Fast Charging
                          ...                    
1010      5000 mAh Battery with 10W Fast Charging
1016    5000 mAh Battery with 22.5W Fast Charging
1017      5080 mAh Battery with 67W Fast Charging
1018          5000 mAh Battery with Fast Charging
1019          5000 mAh Battery with Fast Charging
Name: battery, Length: 725, dtype: object

### Display

In [106]:
df.columns

Index(['model', 'price', 'rating', 'sim', 'processor', 'ram', 'battery',
       'display', 'camera', 'card', 'os'],
      dtype='object')

In [107]:
final_df["display"]

0       6.7 inches, 1440 x 3216 px, 120 Hz Display wit...
1       6.59 inches, 1080 x 2412 px, 120 Hz Display wi...
2       6.6 inches, 1080 x 2408 px, 90 Hz Display with...
3       6.55 inches, 1080 x 2400 px, 120 Hz Display wi...
4       6.7 inches, 1080 x 2412 px, 120 Hz Display wit...
                              ...                        
1010    6.5 inches, 720 x 1600 px Display with Water D...
1016    6.5 inches, 720 x 1600 px Display with Water D...
1017    6.6 inches, 1080 x 2460 px, 144 Hz Display wit...
1018    6.8 inches, 1080 x 2400 px Display with Punch ...
1019    6.5 inches, 1080 x 2400 px Display with Water ...
Name: display, Length: 725, dtype: object

In [108]:
def clean_display(ser: pd.Series):
    temp_df = pd.DataFrame()
    return (
        temp_df
        .assign(
            display_size= ser.str.extract("(\d+\.?\d*)\sinches").astype(np.float64),
            resolution= ser.str.extract("(\d+.?x.?\d+).?(?:px)?"),
            refresh_rate= ser.str.extract("(\d+)\sHz")
        )
    )

<>:6: SyntaxWarning: invalid escape sequence '\d'
<>:7: SyntaxWarning: invalid escape sequence '\d'
<>:8: SyntaxWarning: invalid escape sequence '\d'
<>:6: SyntaxWarning: invalid escape sequence '\d'
<>:7: SyntaxWarning: invalid escape sequence '\d'
<>:8: SyntaxWarning: invalid escape sequence '\d'
C:\Users\dipak\AppData\Local\Temp\ipykernel_18756\970162781.py:6: SyntaxWarning: invalid escape sequence '\d'
  display_size= ser.str.extract("(\d+\.?\d*)\sinches").astype(np.float64),
C:\Users\dipak\AppData\Local\Temp\ipykernel_18756\970162781.py:7: SyntaxWarning: invalid escape sequence '\d'
  resolution= ser.str.extract("(\d+.?x.?\d+).?(?:px)?"),
C:\Users\dipak\AppData\Local\Temp\ipykernel_18756\970162781.py:8: SyntaxWarning: invalid escape sequence '\d'
  refresh_rate= ser.str.extract("(\d+)\sHz")


In [109]:
clean_display(final_df["display"]).join(final_df["display"])

,display_size,resolution,refresh_rate,display
0,6.70,1440 x 3216,120,"6.7 inches, 1440 x 3216 px, 120 Hz Display wit..."
1,6.59,1080 x 2412,120,"6.59 inches, 1080 x 2412 px, 120 Hz Display wi..."
2,6.60,1080 x 2408,90,"6.6 inches, 1080 x 2408 px, 90 Hz Display with..."
3,6.55,1080 x 2400,120,"6.55 inches, 1080 x 2400 px, 120 Hz Display wi..."
4,6.70,1080 x 2412,120,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit..."
...,...,...,...,...
1010,6.50,720 x 1600,NaN,"6.5 inches, 720 x 1600 px Display with Water D..."
1016,6.50,720 x 1600,NaN,"6.5 inches, 720 x 1600 px Display with Water D..."
1017,6.60,1080 x 2460,144,"6.6 inches, 1080 x 2460 px, 144 Hz Display wit..."
1018,6.80,1080 x 2400,NaN,"6.8 inches, 1080 x 2400 px Display with Punch ..."


In [110]:
clean_display(final_df["display"]).isna().sum()

display_size      0
resolution        0
refresh_rate    327
dtype: int64

In [111]:
df.columns

Index(['model', 'price', 'rating', 'sim', 'processor', 'ram', 'battery',
       'display', 'camera', 'card', 'os'],
      dtype='object')

### Camera

In [113]:
final_df["camera"]

0       50 MP + 48 MP + 32 MP Triple Rear & 16 MP Fron...
1       64 MP + 2 MP + 2 MP Triple Rear & 16 MP Front ...
2       50 MP + 2 MP + 2 MP Triple Rear & 13 MP Front ...
3       50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...
4       108 MP + 8 MP + 2 MP Triple Rear & 16 MP Front...
                              ...                        
1010                        8 MP Rear & 5 MP Front Camera
1016    48 MP + 2 MP + Depth Sensor Triple Rear & 8 MP...
1017    64 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...
1018    108 MP + 8 MP + 2 MP Triple Rear & 32 MP Front...
1019    64 MP + 8 MP + 5 MP Triple Rear & 32 MP Front ...
Name: camera, Length: 725, dtype: object

In [114]:
(
    final_df["camera"]
    .str.replace("\u2009","")
    .str.strip()
    .str.extract("(\d+)MP.* (\w+) Rear.* (\d+)MP Front")
    .rename(columns={k:v for k,v in enumerate(["rear_camera", "num_of_cameras", "front_camera"])})
)

<>:5: SyntaxWarning: invalid escape sequence '\d'
<>:5: SyntaxWarning: invalid escape sequence '\d'
C:\Users\dipak\AppData\Local\Temp\ipykernel_18756\4267253823.py:5: SyntaxWarning: invalid escape sequence '\d'
  .str.extract("(\d+)MP.* (\w+) Rear.* (\d+)MP Front")


,rear_camera,num_of_cameras,front_camera
0,50,Triple,16
1,64,Triple,16
2,50,Triple,13
3,50,Triple,16
4,108,Triple,16
...,...,...,...
1010,NaN,NaN,NaN
1016,48,Triple,8
1017,64,Triple,16
1018,108,Triple,32


In [115]:
def clean_camera(ser: pd.Series):
    temp_df = pd.DataFrame()

    return(
            temp_df
            .assign(
                rear_camera= ser.str.extract("(\d+).?MP").astype(np.float64),
                front_camera=ser.str.extract("(\d+).?MP\sFront").astype(np.float64),
                num_of_cameras=ser.str.extract("(\w{4,6})\sRear"),
            )
    )

<>:7: SyntaxWarning: invalid escape sequence '\d'
<>:8: SyntaxWarning: invalid escape sequence '\d'
<>:9: SyntaxWarning: invalid escape sequence '\w'
<>:7: SyntaxWarning: invalid escape sequence '\d'
<>:8: SyntaxWarning: invalid escape sequence '\d'
<>:9: SyntaxWarning: invalid escape sequence '\w'
C:\Users\dipak\AppData\Local\Temp\ipykernel_18756\918589767.py:7: SyntaxWarning: invalid escape sequence '\d'
  rear_camera= ser.str.extract("(\d+).?MP").astype(np.float64),
C:\Users\dipak\AppData\Local\Temp\ipykernel_18756\918589767.py:8: SyntaxWarning: invalid escape sequence '\d'
  front_camera=ser.str.extract("(\d+).?MP\sFront").astype(np.float64),
C:\Users\dipak\AppData\Local\Temp\ipykernel_18756\918589767.py:9: SyntaxWarning: invalid escape sequence '\w'
  num_of_cameras=ser.str.extract("(\w{4,6})\sRear"),


In [116]:
clean_camera(final_df["camera"]).join(final_df["camera"])

,rear_camera,front_camera,num_of_cameras,camera
0,50.0,16.0,Triple,50 MP + 48 MP + 32 MP Triple Rear & 16 MP Fron...
1,64.0,16.0,Triple,64 MP + 2 MP + 2 MP Triple Rear & 16 MP Front ...
2,50.0,13.0,Triple,50 MP + 2 MP + 2 MP Triple Rear & 13 MP Front ...
3,50.0,16.0,Triple,50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...
4,108.0,16.0,Triple,108 MP + 8 MP + 2 MP Triple Rear & 16 MP Front...
...,...,...,...,...
1010,8.0,5.0,NaN,8 MP Rear & 5 MP Front Camera
1016,48.0,8.0,Triple,48 MP + 2 MP + Depth Sensor Triple Rear & 8 MP...
1017,64.0,16.0,Triple,64 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...
1018,108.0,32.0,Triple,108 MP + 8 MP + 2 MP Triple Rear & 32 MP Front...


In [117]:
clean_camera(final_df["camera"]).isna().sum()

rear_camera        0
front_camera      17
num_of_cameras    53
dtype: int64

In [118]:
clean_camera(final_df["camera"]).loc[clean_camera(final_df["camera"]).isna().any(axis=1)].join(final_df["camera"])

,rear_camera,front_camera,num_of_cameras,camera
52,13.0,8.0,NaN,13 MP Rear & 8 MP Front Camera
70,8.0,5.0,NaN,8 MP Rear & 5 MP Front Camera
71,64.0,NaN,Triple,64 MP + 8 MP + 2 MP Triple Rear & 50 MP + 8 MP...
79,13.0,5.0,NaN,13 MP Rear & 5 MP Front Camera
125,64.0,NaN,Triple,64 MP + 8 MP + 2 MP Triple Rear & 50 MP + 8 MP...
...,...,...,...,...
979,13.0,5.0,NaN,13 MP Rear & 5 MP Front Camera
990,40.0,NaN,Quad,40 MP Quad Rear & 32 MP Dual Front Camera
996,13.0,5.0,NaN,13 MP Rear & 5 MP Front Camera
1005,20.0,13.0,NaN,20 MP Rear & 13 MP Front Camera


In [119]:
df.columns

Index(['model', 'price', 'rating', 'sim', 'processor', 'ram', 'battery',
       'display', 'camera', 'card', 'os'],
      dtype='object')

### Card

In [121]:
(
    final_df["card"].sample(20)
)

466     Memory Card Supported, upto 512 GB
981     Memory Card Supported, upto 256 GB
650     Memory Card Supported, upto 256 GB
799     Memory Card Supported, upto 256 GB
122              Memory Card Not Supported
432       Memory Card Supported, upto 1 TB
793     Memory Card Supported, upto 256 GB
982     Memory Card Supported, upto 256 GB
426              Memory Card Not Supported
356       Memory Card Supported, upto 1 TB
441                   Memory Card (Hybrid)
955              Memory Card Not Supported
693       Memory Card Supported, upto 1 TB
488     Memory Card Supported, upto 256 GB
597                  Memory Card Supported
855     Memory Card Supported, upto 256 GB
655     Memory Card Supported, upto 512 GB
964                  Memory Card Supported
879       Memory Card Supported, upto 1 TB
1001      Memory Card Supported, upto 1 TB
Name: card, dtype: object

In [122]:
def clean_card(ser: pd.Series):
    temp_df = pd.DataFrame()

    return (
        temp_df
        .assign(
            card_size=lambda df_: ser.str.extract("(\d+).?(?:TB|GB)").astype(np.float64).replace(1.0, 1024.0),
            has_card=lambda df_: np.where(df_["card_size"].notna(), True, False)
        )
    )

<>:7: SyntaxWarning: invalid escape sequence '\d'
<>:7: SyntaxWarning: invalid escape sequence '\d'
C:\Users\dipak\AppData\Local\Temp\ipykernel_18756\595637481.py:7: SyntaxWarning: invalid escape sequence '\d'
  card_size=lambda df_: ser.str.extract("(\d+).?(?:TB|GB)").astype(np.float64).replace(1.0, 1024.0),


In [123]:
clean_card(final_df["card"])

,card_size,has_card
0,NaN,False
1,1024.0,True
2,1024.0,True
3,1024.0,True
4,NaN,False
...,...,...
1010,1024.0,True
1016,1024.0,True
1017,NaN,False
1018,1024.0,True


### Os

In [125]:
final_df["os"].unique()

array(['Android v13', 'Android v12', 'iOS v16', 'iOS v15', 'Android v11',
       'Android v10', 'iOS v13', 'HarmonyOS v2', 'iOS v15.0',
       'Android v10.0', 'Android v8.1 (Oreo)', 'iOS v13.0',
       'Android v11.0', 'Android v12.1', 'iOS v12.3',
       'Android v9.0 (Pie)', 'iOS v17', 'Android v4.4.2 (KitKat)',
       'Hongmeng OS v3.0', 'Android v8.0 (Oreo)',
       'Pragati OS (Powered by Android)', 'Harmony v2.0',
       'Hongmeng OS v4.0', 'HarmonyOS v2.0', 'Android v6.0 (Marshmallow)',
       'EMUI v12', 'Android v7.1 (Nougat)', 'Android', 'Android v9 (Pie)'],
      dtype=object)

## Semi Final step

In [127]:
df.columns

Index(['model', 'price', 'rating', 'sim', 'processor', 'ram', 'battery',
       'display', 'camera', 'card', 'os'],
      dtype='object')

In [128]:
final_df.columns[3:10].to_list() + ["model"]

['sim', 'processor', 'ram', 'battery', 'display', 'camera', 'card', 'model']

In [129]:
def clean_phone_data(data: pd.DataFrame):
    return(
        data
        .join(data["model"].pipe(clean_model))      # cleaning model columnn --> brand, model_name
        .join(data["sim"].pipe(clean_sim))
        .join(data["processor"].pipe(clean_processor))
        .join(data["ram"].pipe(clean_ram))
        .join(data["battery"].pipe(clean_battery))
        .join(data["display"].pipe(clean_display))
        .join(data["camera"].pipe(clean_camera))
        .join(data["card"].pipe(clean_card))
        .drop(columns=["model"] + data.columns[3:10].to_list() + ["os"])
        .assign(
            price=lambda df_ : df_["price"].pipe(clean_price)
        )
    )

In [130]:
final_df.sample(10)

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
802,Vivo V15 Pro (8GB RAM +128GB),"₹25,000",78.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Snapdragon 675, Octa Core, 2 GHz Processor","8 GB RAM, 128 GB inbuilt",3700 mAh Battery with Fast Charging,"6.39 inches, 1080 x 2316 px Display",48 MP + 8 MP + 5 MP Triple Rear & 32 MP Front ...,"Memory Card Supported, upto 256 GB",Android v9.0 (Pie)
289,Infinix Hot 20 (6GB RAM+ 128GB),"₹19,999",78.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Dimensity 810, Octa Core, 2.4 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 18W Fast Charging,"6.6 inches, 1080 x 2408 px, 120 Hz Display wit...",50 MP Dual Rear & 8 MP Front Camera,"Memory Card Supported, upto 1 TB",Android v12
953,Vivo Y55 4G,"₹22,990",83.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Snapdragon 680, Octa Core, 2.4 GHz Processor","8 GB RAM, 128 GB inbuilt",5000 mAh Battery with 44W Fast Charging,"6.44 inches, 1080 x 2400 px, 120 Hz Display wi...",50 MP + 2 MP + 2 MP Triple Rear & 16 MP Front ...,"Memory Card (Hybrid), upto 1 TB",Android v12
440,Samsung Galaxy A52 (8GB RAM + 128GB),"₹22,494",86.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Snapdragon 720G, Octa Core, 2.3 GHz Processor","8 GB RAM, 128 GB inbuilt",4500 mAh Battery with 25W Fast Charging,"6.5 inches, 1080 x 2400 px, 90 Hz Display with...",64 MP Quad Rear & 32 MP Front Camera,"Memory Card Supported, upto 1 TB",Android v11
179,Vivo V25 (12GB RAM + 256GB),"₹31,999",85.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Dimensity 900, Octa Core, 2 GHz Processor","12 GB RAM, 256 GB inbuilt",4500 mAh Battery with 44W Fast Charging,"6.44 inches, 1080 x 2408 px, 90 Hz Display wit...",64 MP + 8 MP + 2 MP Triple Rear & 50 MP Front ...,"Memory Card (Hybrid), upto 1 TB",Android v12
518,Samsung Galaxy M13 (4GB RAM + 64GB),"₹10,630",71.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Exynos 850, Octa Core, 2 GHz Processor","4 GB RAM, 64 GB inbuilt",6000 mAh Battery with 15W Fast Charging,"6.6 inches, 1080 x 2400 px Display with Water ...",50 MP + 5 MP + 2 MP Triple Rear & 8 MP Front C...,"Memory Card Supported, upto 1 TB",Android v12
758,iKall Z19,"₹7,999",61.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Quad Core, 1.6 GHz Processor","4 GB RAM, 32 GB inbuilt",5000 mAh Battery,"6.5 inches, 720 x 1600 px Display with Water D...",20 MP Rear & 13 MP Front Camera,"Memory Card (Hybrid), upto 128 GB",Android v11.0
389,Realme 9i (6GB RAM + 128GB),"₹14,999",78.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Snapdragon 680, Octa Core, 2.4 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 33W Fast Charging,"6.6 inches, 1080 x 2412 px, 90 Hz Display with...",50 MP + 2 MP + 2 MP Triple Rear & 16 MP Front ...,"Memory Card Supported, upto 1 TB",Android v11
744,itel S16,"₹5,990",NaN,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Unisoc SC7731E, Quad Core, 1.3 GHz Processor","1 GB RAM, 16 GB inbuilt",4000 mAh Battery,"6.5 inches, 1600 x 720 px Display with Water D...",8 MP + 2 MP + 0.3 MP Triple Rear & 8 MP Front ...,"Memory Card Supported, upto 32 GB",Android v10
260,Poco M4 5G (6GB RAM + 128GB),"₹13,499",79.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster","Dimensity 700 5G, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 18W Fast Charging,"6.58 inches, 1080 x 2400 px, 90 Hz Display wit...",50 MP + 2 MP Dual Rear & 8 MP Front Camera,"Memory Card Supported, upto 512 GB",Android v12


In [131]:
clean_phone_data(final_df)

,price,rating,brand,model_name,number_of_sim,has_5G,has_NFC,has_IR_Blaster,cpu_speed,number_of_cores,...,fast_charging,fast_charging_power,display_size,resolution,refresh_rate,rear_camera,front_camera,num_of_cameras,card_size,has_card
0,54999.0,89.0,Oneplus,11 5G,2,True,True,False,3.20,8,...,True,100,6.70,1440 x 3216,120,50.0,16.0,Triple,NaN,False
1,19989.0,81.0,Oneplus,Nord CE 2 Lite 5G,2,True,False,False,2.20,8,...,True,33,6.59,1080 x 2412,120,64.0,16.0,Triple,1024.0,True
2,16499.0,75.0,Samsung,Galaxy A14 5G,2,True,False,False,2.40,8,...,True,15,6.60,1080 x 2408,90,50.0,13.0,Triple,1024.0,True
3,14999.0,81.0,Motorola,Moto G62 5G,2,True,False,False,2.20,8,...,True,NaN,6.55,1080 x 2400,120,50.0,16.0,Triple,1024.0,True
4,24999.0,82.0,Realme,10 Pro Plus,2,True,False,False,2.60,8,...,True,67,6.70,1080 x 2412,120,108.0,16.0,Triple,NaN,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1010,8720.0,64.0,Realme,Narzo 50i Prime,2,False,False,False,1.82,8,...,True,10,6.50,720 x 1600,NaN,8.0,5.0,NaN,1024.0,True
1016,14990.0,75.0,Honor,X8 5G,2,True,False,False,2.20,8,...,True,5,6.50,720 x 1600,NaN,48.0,8.0,Triple,1024.0,True
1017,28990.0,85.0,Poco,X4 GT 5G,2,True,True,True,2.85,8,...,True,67,6.60,1080 x 2460,144,64.0,16.0,Triple,NaN,False
1018,19990.0,80.0,Motorola,Moto G91 5G,2,True,True,False,2.20,8,...,True,NaN,6.80,1080 x 2400,NaN,108.0,32.0,Triple,1024.0,True


## Left shifted phones

In [133]:
get_left_shifted_phones(df).shape

(285, 11)

In [134]:
(
    get_left_shifted_phones(df)
    .drop(index=[111, 149, 307, 363, 439, 448, 607, 628, 656, 754, 762, 853, 913, 914])
    .sample(50)
#     .loc[:,"card"]
#     .unique()
)

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
415,Xiaomi Mi 11X,"₹23,890",80.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster","Snapdragon 870, Octa Core, 3.2 GHz Processor","6 GB RAM, 128 GB inbuilt",4520 mAh Battery with 33W Fast Charging,"6.67 inches, 1080 x 2400 px, 120 Hz Display wi...",48 MP + 8 MP + 5 MP Triple Rear & 20 MP Front ...,Android v11,No FM Radio
724,Huawei Mate 50 Pro 5G,"₹99,990",85.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Kirin 9000E, Octa Core, 3.13 GHz Processor","8 GB RAM, 256 GB inbuilt",4500 mAh Battery with 66W Fast Charging,"6.76 inches, 1440 x 3200 px, 120 Hz Display wi...",64 MP + 50 MP + 13 MP Triple Rear & 13 MP Fron...,Memory Card Not Supported,Hongmeng OS v4.0
39,OPPO Reno 9 5G,"₹28,499",83.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 778G, Octa Core, 2.4 GHz Processor","8 GB RAM, 256 GB inbuilt",4500 mAh Battery with 67W Fast Charging,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit...",64 MP + 2 MP Dual Rear & 32 MP Front Camera,Android v13,No FM Radio
891,iQOO 9 5G (12GB RAM + 256GB),"₹39,990",85.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 888+, Octa Core, 2.9 GHz Processor","12 GB RAM, 256 GB inbuilt",4350 mAh Battery with 120W Fast Charging,"6.56 inches, 1080 x 2376 px, 120 Hz Display wi...",48 MP + 13 MP + 13 MP Triple Rear & 16 MP Fron...,Android v12,No FM Radio
760,Vivo S17 Pro,"₹49,990",77.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Dimensity 9200, Octa Core Processor","12 GB RAM, 256 GB inbuilt",5000 mAh Battery with 120W Fast Charging,"6.67 inches, 1080 x 2400 px, 120 Hz Display wi...",50 MP + 13 MP + 2 MP Triple Rear & 32 MP Front...,Android v13,No FM Radio
817,itel Magic X,"₹2,239",NaN,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi",No 3G,T117,"48 MB RAM, 128 MB inbuilt",1200 mAh Battery,"2.4 inches, 240 x 320 px Display",1.3 MP Rear Camera,"Memory Card Supported, upto 64 GB"
400,Jio JioPhone 2,"₹2,999",NaN,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Dual Core, 1 GHz Processor","512 MB RAM, 4 GB inbuilt",2000 mAh Battery,"2.4 inches, 320 x 240 px Display",2 MP Rear & 0.3 MP Front Camera,"Memory Card Supported, upto 128 GB",KAI OS
97,iQOO 11 5G,"₹59,999",88.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor","8 GB RAM, 256 GB inbuilt",5000 mAh Battery with 120W Fast Charging,"6.78 inches, 1440 x 3200 px, 144 Hz Display wi...",50 MP + 13 MP + 8 MP Triple Rear & 16 MP Front...,Android v13,No FM Radio
652,Realme X7 Max,"₹26,999",84.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Dimensity 1200, Octa Core, 3 GHz Processor","8 GB RAM, 128 GB inbuilt",4500 mAh Battery with 50W Fast Charging,"6.43 inches, 1080 x 2400 px, 120 Hz Display wi...",64 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,Android v11,No FM Radio
658,Letv Y1 Pro Plus,"₹5,999",60.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Tiger T610, Quad Core, 1.8 GHz Processor","6 GB RAM, 64 GB inbuilt",4000 mAh Battery with 10W Fast Charging,"6.5 inches, 720 x 1560 px Display with Small N...",8 MP + Depth Sensor Dual Rear & 5 MP Front Camera,Android v11,No FM Radio


In [135]:
(
    get_left_shifted_phones(df)
    .drop(index=[111, 149, 307, 363, 439, 448, 607, 628, 656, 754, 762, 853, 913, 914])
    # .loc[:,"card"]
    # .unique()
)

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
8,Nothing Phone 1,"₹26,749",85.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 778G Plus, Octa Core, 2.5 GHz Proce...","8 GB RAM, 128 GB inbuilt",4500 mAh Battery with 33W Fast Charging,"6.55 inches, 1080 x 2400 px, 120 Hz Display wi...",50 MP + 50 MP Dual Rear & 16 MP Front Camera,Android v12,No FM Radio
9,OnePlus Nord 2T 5G,"₹28,999",84.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Dimensity 1300, Octa Core, 3 GHz Processor","8 GB RAM, 128 GB inbuilt",4500 mAh Battery with 80W Fast Charging,"6.43 inches, 1080 x 2400 px, 90 Hz Display wit...",50 MP + 8 MP + 2 MP Triple Rear & 32 MP Front ...,Android v12,No FM Radio
12,Xiaomi Redmi Note 12 Pro 5G,"₹24,762",79.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster","Dimensity 1080, Octa Core, 2.6 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 67W Fast Charging,"6.67 inches, 1080 x 2400 px, 120 Hz Display wi...",50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,Android v12,No FM Radio
17,OPPO Reno 9 Pro Plus,"₹45,999",86.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8+ Gen1, Octa Core, 3.2 GHz Processor","16 GB RAM, 256 GB inbuilt",4700 mAh Battery with 80W Fast Charging,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit...",50 MP + 8 MP + 2 MP Triple Rear & 32 MP Front ...,Android v13,No FM Radio
18,OnePlus 10R 5G,"₹32,999",86.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Dimensity 8100 Max, Octa Core, 2.85 GHz Processor","8 GB RAM, 128 GB inbuilt",5000 mAh Battery with 80W Fast Charging,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit...",50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,Android v12,Bluetooth
...,...,...,...,...,...,...,...,...,...,...,...
1009,Xiaomi Civi 3,"₹32,990",86.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...","Dimensity 8200, Octa Core, 3.1 GHz Processor","8 GB RAM, 256 GB inbuilt",5000 mAh Battery with 80W Fast Charging,"6.7 inches, 1080 x 2400 px, 120 Hz Display wit...",64 MP + 20 MP + 2 MP Triple Rear & 32 MP + 32 ...,Android v13,No FM Radio
1011,Oppo Find X6,"₹69,990",89.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor","8 GB RAM, 256 GB inbuilt",4700 mAh Battery with 120W Fast Charging,"6.73 inches, 1080 x 2400 px, 120 Hz Display wi...",50 MP + 48 MP + 32 MP Triple Rear & 32 MP Fron...,Android v12,No FM Radio
1012,itel A23s,"₹4,787",NaN,"Dual Sim, 3G, 4G, Wi-Fi","Spreadtrum SC9832E, Quad Core, 1.4 GHz Processor","2 GB RAM, 32 GB inbuilt",3020 mAh Battery,"5 inches, 854 x 480 px Display",2 MP Rear Camera,Android v11,No FM Radio
1013,Google Pixel 8 Pro,"₹70,990",80.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Google Tensor 3, Octa Core Processor","12 GB RAM, 256 GB inbuilt",5000 mAh Battery with 67W Fast Charging,"6.73 inches, 1440 x 3120 px, 120 Hz Display wi...",50 MP + 50 MP + 50 MP Triple Rear & 12 MP Fron...,Android v13,No FM Radio


In [136]:
index_of_left_shifted_phones = (
                                get_left_shifted_phones(df_without_feature_phones)
                                .drop(index=[111, 149, 307, 363, 439, 448, 607, 628, 762, 853, 913, 914])
                                .loc[:,"card"]
                                .str.contains(pat="Android|iOS",case=False)
                                .index.to_list()
                            )

### **Apple data** --> left shifted phones

In [138]:
apple_no_shifting = get_left_shifted_phones(get_apple_phones(df)).loc[:, :"ram"]

In [139]:
apple_shifting = get_left_shifted_phones(get_apple_phones(df)).loc[:, "battery":].shift(periods=1, axis=1)

In [140]:
apple_correct_data = apple_no_shifting.join(apple_shifting)

apple_correct_data

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
111,Apple iPhone 12,"₹51,999",74.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 64 GB inbuilt",None,"6.1 inches, 1170 x 2532 px Display with Large ...",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14
149,Apple iPhone 12 Mini,"₹40,999",74.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 64 GB inbuilt",None,"5.4 inches, 1080 x 2340 px Display",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14
307,Apple iPhone 12 (128GB),"₹55,999",75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 128 GB inbuilt",None,"6.1 inches, 1170 x 2532 px Display with Large ...",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14
363,Apple iPhone 12 Mini (128GB),"₹45,999",75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 128 GB inbuilt",None,"5.4 inches, 1080 x 2340 px Display",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14
439,Apple iPhone SE 3 2022,"₹43,900",NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A15, Hexa Core, 3.22 GHz Processor",64 GB inbuilt,None,"4.7 inches, 750 x 1334 px Display",12 MP Rear & 7 MP Front Camera,Memory Card Not Supported,iOS v15
448,Apple iPhone 15 Pro,"₹1,30,990",75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC",Bionic A16,"8 GB RAM, 128 GB inbuilt",None,"6.06 inches, 1170 x 2532 px, 120 Hz Display wi...",50 MP + 12 MP + 12 MP Triple Rear & 12 MP Fron...,Memory Card Not Supported,iOS v15
607,Apple iPhone 7s,"₹52,990",NaN,"Single Sim, 3G, 4G, VoLTE, Wi-Fi","Fusion APL1024, Quad Core, 2.37 GHz Processor","3 GB RAM, 32 GB inbuilt",None,2230 mAh Battery,"4.7 inches, 750 x 1334 px Display",13 MP Rear & 7 MP Front Camera,iOS v10
628,Apple iPhone 12 Pro (512GB),"₹1,39,900",80.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","6 GB RAM, 512 GB inbuilt",None,"6.1 inches, 1170 x 2532 px Display with Large ...",12 MP + 12 MP + 12 MP Triple Rear & 12 MP Fron...,Memory Card Not Supported,iOS v14.0
656,Apple iPhone SE 2020,"₹39,900",63.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi, NFC","A13 Bionic, Hexa Core, 2.65 GHz Processor","3 GB RAM, 64 GB inbuilt",None,1821 mAh Battery with 18W Fast Charging,"4.7 inches, 750 x 1334 px Display",12 MP Rear & 7 MP Front Camera,iOS v13
754,Apple iPod Touch (7th Gen),"₹18,900",NaN,Wi-Fi,32 GB inbuilt,"4 inches, 640 x 1136 px Display",None,8 MP Rear & 1.2 MP Front Camera,iOS v12,No FM Radio,Bluetooth


In [141]:
apple_correct_data.isna().sum()

model         0
price         0
rating        3
sim           0
processor     0
ram           0
battery      14
display       0
camera        0
card          0
os            0
dtype: int64

# Final Cleaning Step

In [143]:
725 + 14

739

In [144]:
df_shifted_phones = df_without_feature_phones.loc[index_of_left_shifted_phones,:]

In [145]:
final_data = pd.concat([final_df, apple_correct_data, df_shifted_phones])

final_data.duplicated().sum()

# drop duplicates

final_data.drop_duplicates(inplace=True)

final_data.shape

(966, 11)

In [146]:
final_data.duplicated().sum()

0

In [147]:
len(feature_phones_index)

43

In [148]:
1020 - 43

977

In [149]:
def clean_phone_data(data: pd.DataFrame):
    return(
        data
        .join(data["model"].pipe(clean_model))
        .join(data["sim"].pipe(clean_sim))
        .join(data["processor"].pipe(clean_processor))
        .join(data["ram"].pipe(clean_ram))
        .join(data["battery"].pipe(clean_battery))
        .join(data["display"].pipe(clean_display))
        .join(data["camera"].pipe(clean_camera))
        .join(data["card"].pipe(clean_card))
        .drop(columns=["model"] + data.columns[3:10].to_list() + ["os"])
        .assign(
            price=lambda df_ : df_["price"].pipe(clean_price)
        )
    )

In [150]:
cleaned_data = clean_phone_data(final_data)

cleaned_data.head()

,price,rating,brand,model_name,number_of_sim,has_5G,has_NFC,has_IR_Blaster,cpu_speed,number_of_cores,...,fast_charging,fast_charging_power,display_size,resolution,refresh_rate,rear_camera,front_camera,num_of_cameras,card_size,has_card
0,54999.0,89.0,Oneplus,11 5G,2,True,True,False,3.2,8,...,True,100,6.70,1440 x 3216,120,50.0,16.0,Triple,NaN,False
1,19989.0,81.0,Oneplus,Nord CE 2 Lite 5G,2,True,False,False,2.2,8,...,True,33,6.59,1080 x 2412,120,64.0,16.0,Triple,1024.0,True
2,16499.0,75.0,Samsung,Galaxy A14 5G,2,True,False,False,2.4,8,...,True,15,6.60,1080 x 2408,90,50.0,13.0,Triple,1024.0,True
3,14999.0,81.0,Motorola,Moto G62 5G,2,True,False,False,2.2,8,...,True,NaN,6.55,1080 x 2400,120,50.0,16.0,Triple,1024.0,True
4,24999.0,82.0,Realme,10 Pro Plus,2,True,False,False,2.6,8,...,True,67,6.70,1080 x 2412,120,108.0,16.0,Triple,NaN,False


In [151]:
# save as csv

cleaned_data.to_csv("cleaned_smartphones_data.csv", index=False)

In [152]:
# check for missing values in clean data

cleaned_data.isna().sum()

price                    0
rating                  91
brand                    0
model_name               0
number_of_sim            0
has_5G                   0
has_NFC                  0
has_IR_Blaster           0
cpu_speed               47
number_of_cores          0
processor_name          58
memory                   8
storage                  5
capacity                18
fast_charging           14
fast_charging_power    210
display_size             7
resolution               7
refresh_rate           365
rear_camera             20
front_camera            46
num_of_cameras          77
card_size              470
has_card                 0
dtype: int64

In [153]:
# shape of cleaned data

cleaned_data.shape

(966, 24)